{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure. Functional Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import cPickle\n",
    "import glob\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "\n",
    "import cdpybio as cpb\n",
    "from ipyparallel import Client\n",
    "import matplotlib as mpl\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pybedtools as pbt\n",
    "from scipy.stats import fisher_exact\n",
    "import seaborn as sns\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import statsmodels as sms\n",
    "import tabix\n",
    "import vcf as pyvcf\n",
    "import weblogolib as logo\n",
    "\n",
    "import cardipspy as cpy\n",
    "import ciepy\n",
    "\n",
    "%matplotlib inline\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "dy_name = 'figure_functional_annotation_and_power_analysis'\n",
    "\n",
    "import socket\n",
    "if socket.gethostname() == 'fl-hn1' or socket.gethostname() == 'fl-hn2':\n",
    "    dy = os.path.join(ciepy.root, 'sandbox', 'tmp', dy_name)\n",
    "    cpy.makedir(dy)\n",
    "    pbt.set_tempdir(dy)\n",
    "    \n",
    "outdir = os.path.join(ciepy.root, 'output', dy_name)\n",
    "cpy.makedir(outdir)\n",
    "\n",
    "private_outdir = os.path.join(ciepy.root, 'private_output', dy_name)\n",
    "cpy.makedir(private_outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sns.set_style('whitegrid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fn = os.path.join(ciepy.root, 'output', 'functional_annotation_analysis', 'encode_dnase_res.tsv')\n",
    "dnase_res = pd.read_table(fn, index_col=0)\n",
    "\n",
    "fn = os.path.join(ciepy.root, 'output', 'functional_annotation_analysis', 'encode_stem_cell_chip_seq_res.tsv')\n",
    "tf_res = pd.read_table(fn, index_col=0)\n",
    "tf_res = tf_res.drop_duplicates(subset=['target'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fn = os.path.join(ciepy.root, 'output', 'gtex_analysis', 'plot_data.tsv')\n",
    "plotd = pd.read_table(fn, index_col=0)\n",
    "plotd = plotd.drop('Testis')\n",
    "fn = os.path.join(ciepy.root, 'output', 'input_data', 'rnaseq_metadata.tsv')\n",
    "rna_meta = pd.read_table(fn, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "123 samples from 78 cell types.\n",
      "9 skin fibroblasts.\n"
     ]
    }
   ],
   "source": [
    "tdf = dnase_res[dnase_res.biosample_type != 'immortalized cell line'].copy(deep=True)\n",
    "a = tdf.shape[0]\n",
    "b = tdf.cell_type.value_counts().shape[0]\n",
    "print('{} samples from {} cell types.'.format(a, b))\n",
    "print('{} skin fibroblasts.'.format(sum(tdf.cell_type == 'skin fibroblast')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tdf.sort_values(by='pvalue', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([16, 27, 39, 61, 62, 64, 72, 90, 94]),)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.where(tdf.cell_type == 'skin fibroblast')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "transcription factor                                                                                39\n",
       "chromatin remodeller, transcription factor                                                           8\n",
       "RNA binding protein, transcription factor                                                            3\n",
       "broad histone mark, histone, histone modification                                                    2\n",
       "chromatin remodeller, other context, other post-translational modification, transcription factor     1\n",
       "histone, histone modification, narrow histone mark                                                   1\n",
       "Name: target_type, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_res.target_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tf_res = tf_res[tf_res.target_type != 'histone, histone modification']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tf_res.sort_values(by='pvalue', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([17]),)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.where(tf_res.target == 'POU5F1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/frazer01/home/cdeboever/software/anaconda/envs/cie/lib/python2.7/site-packages/matplotlib/gridspec.py:302: UserWarning: This figure includes Axes that are not compatible with tight_layout, so its results might be incorrect.\n",
      "  warnings.warn(\"This figure includes Axes that are not \"\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'grasp_res' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-17-30e8217ec9b5>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     63\u001b[0m \u001b[0max2\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mfig\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0madd_subplot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mgs\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     64\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 65\u001b[1;33m \u001b[0mgrasp_res\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mby\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'pvalue'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     66\u001b[0m \u001b[0mt\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mgrasp_res\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mgrasp_res\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpvalue\u001b[0m \u001b[1;33m<\u001b[0m \u001b[1;36m0.05\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     67\u001b[0m \u001b[0mt\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mgrasp_res\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mgrasp_res\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbh_sig\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'grasp_res' is not defined"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeYAAAGTCAYAAAD9fLG9AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XtwzXf+x/HXSSJuJ5qm6lrELUU6QqKMatBK1LbpuKYN\nGjUo2qEMOmx/LklbEl2XXk6tqp1QVWkNvaVUa0N36pqeOjTL0lWbompLEQkS5Pv7wzi7qdzOOnE+\niedjxsj53s47byavfL7f8/18bZZlWQIAAEbw83UBAADgPwhmAAAMQjADAGAQghkAAIMQzAAAGIRg\nBgDAIAHlbZCYmKisrKwS19WvX1/ffPONPvroI7344ovasWOHgoODvV6kN5w7d06LFy/W3/72N509\ne1ZhYWGaNGmSunfv7uvSAABws5V3H3NiYqKKioo0ffr0G9bVqFFD7du315kzZ3T06FHdd9998vMz\ncxA+YsQI/fTTT5o8ebIaNGigdevWaePGjVqzZo0iIiJ8XR4AAJIqMGKWpKCgIHXs2LHU9Xfeeafu\nvPNOrxXlbd9//72ysrK0YsUKdevWTZLUvXt3HTp0SCtWrNDixYt9XCEAANd4ZXi7fv16tWvXTmfP\nnpUkWZYlh8OhXr16qVOnTnr++ee1YsUKtWvXzr3Pww8/rFdeeaXYcebOnauHH37Y/bpdu3Z6++23\nFRcXp86dO2vjxo2SpOzsbD399NPq1KmTunfvrldeeUWXLl0q/Zv081N8fLw6d+7sXmaz2RQaGqpj\nx455owUAAHhFhUbMknT16tUblvn7+0u6FnI2m829fPHixUpLS9PEiRPVvn17rV27VosWLSq2TUl+\nfxxJWrp0qV588UXdcccd6tKliw4fPqzExERFRkbq9ddf1+nTp7VgwQIdO3ZMS5cuLfG44eHheuml\nl4oty8vLU1ZWlnr37l2Rbx8AgFuiQsG8detWhYeHF1tms9lK/LBXfn6+VqxYofHjx2vs2LGSpOjo\naPXv31+HDh3yuMAePXooPj7e/fqVV17R3XffrWXLlrl/MWjRooWGDx+ub7/9Vl26dKnQcZOTk5WX\nl6eRI0d6XBMAAJWlQsHcpUsXvfjii/r958Tq1at3w7Yul0uXL19Wnz59ii3v27fv/xTMoaGhxV7v\n3r1bMTExkv4zio+IiJDdbtfOnTsrFMzJycnKyMjQzJkzi51eBwDA1yoUzHa7XR06dKjQAa9fZw4J\nCSm2vH79+h6Wds1dd911w/E/+OADpaenF1tus9n073//u8xjXb58WS+88II2bdqkadOmafjw4f9T\nTQAAVJYKX2OuqIYNG0qSfvvtNzVo0MC9/PTp08W2s9lsKioqKrbswoUL5R7fbrcrJiZGw4YNu2EE\nX9YnwwsKCjRu3DhlZWUpOTlZTzzxRLnvBQDAreb1m47btWunOnXqKDMzs9jy37+22+3FRriWZWnP\nnj3lHj8qKko//vijOnTooPDwcIWHh6thw4ZasGCBfvjhh1L3mzp1qr799lstWrSIUAYAGMvrI2a7\n3a6nn35aS5cudU9A8sknn2j//v3FPnHds2dPpaWl6b333lPr1q31wQcf6PTp06pbt26Zx3/uuec0\ndOhQTZo0SYMHD1ZBQYGWLFmikydPqn379iXu89VXX2nz5s0aOHCgGjVqpL1797rX1apVS/fee693\nvnkAAG5ShYK5vNucfm/ChAm6evWqVq5cqfz8fPXu3VvDhg3Txx9/7N5m/PjxOnXqlF577TX5+/ur\nf//+Gj9+vFatWlXsfX//3uHh4Vq5cqUWL16sSZMmKTAwUFFRUVqwYEGxU+f/LTMzUzabTR9//HGx\nGiSpTZs2+uyzzzz6/gAAqCzlTsnpqcuXL2vDhg2Kjo4u9gGwqVOn6siRI1q/fr033w4AgGrF66ey\na9SooSVLlmjt2rUaM2aMateure3bt+uLL77Q3Llzvf12AABUK14fMUtSTk6OFixYIKfTqQsXLqhl\ny5YaOXKk+vfv7+23AgCgWqmUYAYAAP8bM5/RCADAbYpgBgDAIAQzAAAGIZgBADAIwQwAgEEIZgAA\nDEIwAwBgEIIZAACDEMwAABiEYAYAwCAEMwAABiGYAQAwCMEMAIBBjA7mvXv3KjEx8YblmZmZGjJk\niBISErR27VofVAYAQOUI8HUBpVm+fLk++eQT1a1bt9jyK1euKDU1VevXr1fNmjU1dOhQ9enTRyEh\nIT6qFAAA7zF2xNyiRQu99dZbNyw/fPiwWrRoIbvdrho1aigqKkpZWVk+qBAAAO8zdsQcGxur48eP\n37A8Ly9PQUFB7td169bV+fPnyz2e0+n0an0AAPNFRUX5ugSPGRvMpbHb7crLy3O/zs/PV7169Sq0\nb1X8B/IVp9NJvzxEzzxDvzxHzzxTVQdkxp7Kvs6yrGKvW7durZycHOXm5qqwsFBZWVnq1KmTj6oD\nAMC7jB8x22w2SVJGRoYuXryo+Ph4/fGPf9SoUaNkWZbi4+PVoEEDH1cJAIB3GB3MTZs2VXp6uiQp\nLi7Ovbx3797q3bu3j6oCAKDyGH8qGwCA2wnBDACAQQhmAAAMQjADAGAQghkAAIMQzAAAGIRgBgDA\nIAQzAAAGIZgBADAIwQwAgEEIZgAADEIwAwBgEIIZAACDEMwAABiEYAYAwCAEMwAABiGYAQAwCMEM\nAIBBCGYAAAxCMAMAYBCCGQAAgxDMAAAYhGAGAMAgBDMAAAYhmAEAMAjBDACAQQhmAAAMQjADAGAQ\nghkAAIMQzAAAGIRgBgDAIAQzAAAGIZgBADAIwQwAgEEIZgAADEIwAwBgEIIZAACDEMwAABiEYAYA\nwCAEMwAABiGYAQAwCMEMAIBBCGYAAAxCMAMAYBCCGQAAgxDMAAAYhGAGAMAgBDMAAAYhmAEAMAjB\nDACAQQhmAAAMQjADAGAQghkAAIMQzAAAGIRgBgDAIAQzAAAGIZgBADAIwQwAgEEIZgAADEIwAwBg\nECOD2bIszZkzRwkJCRoxYoSOHj1abP2nn36qQYMGKT4+XmvWrPFRlQAAeF+ArwsoyebNm1VYWKj0\n9HTt3btXKSkpWrJkiXv9q6++qo0bN6pWrVp67LHHFBcXp6CgIB9WDACAdxgZzE6nU9HR0ZKkiIgI\nZWdnF1vfrl07nTt3TjabTZLcfwMAUNUZGcx5eXnFRsABAQEqKiqSn9+1M+9t27bV4MGDVadOHcXG\nxsput1fouE6ns1Lqra7ol+fomWfol+foWfVnZDDb7Xbl5+e7X/93KB88eFBbt25VZmam6tSpo2nT\npmnTpk165JFHyj1uVFRUpdVc3TidTvrlIXrmGfrlOXrmmar6S4yRH/6KjIzU119/LUlyuVwKCwtz\nrwsKClLt2rUVGBgom82mkJAQ5ebm+qpUAAC8ysgRc2xsrLZt26aEhARJUkpKijIyMnTx4kXFx8fr\niSee0LBhwxQYGKjmzZtr4MCBPq4YAADvMDKYbTabkpOTiy1r2bKl++uEhAR3aAMAUJ0YeSobAIDb\nFcEMAIBBCGYAAAxCMAMAYBCCGQAAgxDMAAAYhGAGAMAgBDMAAAYhmAEAMAjBDACAQQhmAAAMQjAD\nAGAQghkAAIMQzAAAGIRgBgDAIAQzAAAGIZgBADAIwQwAgEEIZgAADEIwAwBgEIIZAACDEMwAABiE\nYAYAwCAEMwAABiGYAQAwCMEMAIBBCGYAAAxCMAMAYBCCGQAAgxDMAAAYhGAGAMAgBDMAAAYhmAEA\nMAjBDACAQQhmAAAMQjADAGAQghkAAIMQzAAAGIRgBgDAIAQzAAAGIZgBADAIwQwAgEEIZgAADEIw\nAwBgEIIZAACDEMwAABiEYAYAwCAEMwAABiGYAQAwCMEMAIBBCGYAAAxCMAMAYBCCGQAAgxDMAAAY\nhGAGAMAgBDMAAAYhmAEAMAjBDACAQQhmAAAMQjADAGCQAF8XUBLLspSUlKSDBw8qMDBQc+fOVbNm\nzdzr9+3bp/nz50uS6tevrz/96U8KDAz0VbkAAHiNkSPmzZs3q7CwUOnp6Zo6dapSUlKKrZ89e7ZS\nU1O1evVqRUdH6+eff/ZRpQAAeJeRI2an06no6GhJUkREhLKzs93rjhw5ouDgYKWlpemHH35Q7969\nFRoa6qNKAQDwLiNHzHl5eQoKCnK/DggIUFFRkSTpzJkzcrlcSkxMVFpamrZv365du3b5qlQAALzK\nyBGz3W5Xfn6++3VRUZH8/K79DhEcHKzmzZurZcuWkqTo6GhlZ2erW7du5R7X6XRWTsHVFP3yHD3z\nDP3yHD2r/owM5sjISG3ZskX9+vWTy+VSWFiYe12zZs104cIFHT16VM2aNZPT6dSQIUMqdNyoqKjK\nKrnacTqd9MtD9Mwz9Mtz9MwzVfWXGCODOTY2Vtu2bVNCQoIkKSUlRRkZGbp48aLi4+M1d+5cTZky\nRZLUuXNn9erVy5flAgDgNUYGs81mU3JycrFl109dS1K3bt20du3aW10WAACVzsgPfwEAcLsimAEA\nMAjBDACAQQhmAAAMQjADAGAQghkAAIMYebsUAADeYOIkI+VNEkMwAwCqrdJCsKKzqFXGduUhmAEA\n1dahQ4dKXJ6Tk1PsYUmlqeh2V69e9bi20hDMAIBq68033yx13ZYtWyp0jPK2O3v2rAYOHKiuXbt6\nVFtpCGYAQLVVv359X5fgMT6VDQCAQRgxAwCqrVOnTlX6e5w9e9arxyOYAQDV1sSJE0tcnp2drfvu\nu6/c/Su6nTfDmWAGAFRbYWFhJS4/f/58qev+l+28eb8015gBADAIwQwAgEEIZgAADEIwAwBgEIIZ\nAACDEMwAABiEYAYAwCAEMwAABiGYAQAwCMEMAIBBCGYAAAxCMAMAYJAKB/O6det0+PDhMrcZM2bM\nTRfkcDi0b9++UtcfP35cc+bMKfMYBw4c0GeffXbTtQAAcKtV+OlSJ06cUJs2bTR+/Hh17NhRx44d\nU0xMjNq3b6958+apSZMmOnr0qKRrAb18+XJt2LBBBQUFCg8PV1pamizLUo8ePdS2bVulp6fLZrOp\ncePGGjVqlGbNmqV69erJ5XKpZ8+e7vd1OBw6efKkmjRpotq1ays2NlaS9NFHH6lmzZp69NFH3e+X\nkJCgVq1aqUOHDrpw4UKxfbt27erl1gEA4H0eP/axoKBAY8eO1alTp/TGG2/I5XJp4sSJCgsL0/79\n+yVJNput2D7Lli1TUlKS7Ha79u/fr2XLlqlx48by9/eXy+XSzp07FR4erqeeekoLFy684T0HDhyo\nyMhITZ48WQ899NAN7/HfX8+bN0+7d++Wy+Uqtq83H8kFAEBl8TiYAwIC3H8sy5Kfn5+KiookSf7+\n/pLkfn3mzBnVqVNHV65cce9//PhxFRUVaejQobrnnnv04Ycfyt/fX5ZlSZL8/G48u379eFevXnWv\n9/f31+XLl93vI0l2u73UfQEAqAo8DubfGzp0qFJTU9WwYUOdPn1aktSkSRMlJycrNzdX0dHRGj16\ntF5++WX5+fnpwQcf1Lhx45SamqqQkBCFhoaqa9eu2rhxo+bPn699+/apT58+xd5j1apV+vzzz9Wj\nRw/5+fnJZrOpS5cumjlzprKzs93hW9Io+vq+cXFxN/utAgBQ6WzW9aGqoRwOh3r27KmOHTve1L5O\np1NRUVGVUGH1RL88R888Q788R888U1a/KtpLX2x30yPmyjZhwgSf7AsAgC9wHzMAAAYhmAEAMAgT\njAAAYBAmGAEAwCBMMAIAgEGYYAQAAIMwwQgAAAZhghGUiH55jp55hn55jp55hglGKgkTjAAAbifc\nxwwAgEEIZgAADMIEIwAAGIQJRgAAMAgTjAAAYBAmGAEAwCBMMAIAgEGYYAQlol+eo2eeoV+eo2ee\nYYKRSsIEIwCA2wn3MQMAYBCCGQAAgxDMAAAYhGAGAMAgBDMAAAYhmAEAMAjBDACAQQhmAAAMYmQw\nW5alOXPmKCEhQSNGjHA/ter3Zs+erUWLFt3i6gAAqDxGBvPmzZtVWFio9PR0TZ06VSkpKTdsk56e\nrkOHDvmgOgAAKo+Rwex0OhUdHS1JioiIUHZ2drH1e/bs0ffff6+EhARflAcAQKUxMpjz8vIUFBTk\nfh0QEOB+gtSvv/4qh8Oh2bNny/DnbwAA4LEKP8Ri3bp16tSpk1q3bl3qNmPGjNHy5ctvqiCHw6GL\nFy8qPz/fvayoqMj9HOb09HTt379fzzzzjH799VcVFBSoVatWGjBggHv7AwcO6J///Kcef/zxYsd2\nOp03Vdvthn55jp55hn55jp5VfxUO5hMnTqhNmzYaP368OnbsqGPHjikmJkbt27fXvHnz1KRJE/eH\ntK4H9IYNG1RQUKDw8HClpaXJsiz16NFDbdu2VXp6umw2mxo3bqxRo0Zp1qxZqlevnlwul2JiYvT1\n11+rX79+mjlzpiTpz3/+s2rXrq1Bgwbp1KlT6tSpk/bu3at69eopIyNDAwYMUEJCglq1aqUOHTro\nwoULcjgcOnnypJo0aaKuXbvyuDQP8Hg5z9Ezz9Avz9Ezz1TVX2I8fuxjQUGBxo4dq1OnTumNN96Q\ny+XSxIkTFRYWpv3790uSbDZbsX2WLVumpKQk2e127d+/X8uWLVPjxo3l7+8vl8ulnTt3Kjw8XE89\n9ZQWLlyorl276tixY0pISNAvv/yiF154QZZlKS0tTQ899NAN7/HfX8+bN0+7d++Wy+WSJA0cOFCR\nkZFV9h8IAHB78TiYAwIC3H8sy5Kfn5/7+q+/v78kuV+fOXNGderU0ZUrV9z7Hz9+XEVFRRo6dKju\nueceffjhh/L393dfL/bz85PNZlNycrKka6e2GzZsqC5dumjTpk3uU9r+/v7q3Lmz+vfvryFDhkiS\n7Hb7DfVerwUAgKrA42D+vaFDhyo1NVUNGzbU6dOnJUlNmjRRcnKycnNzFR0drdGjR+vll1+Wn5+f\nHnzwQY0bN06pqakKCQlRaGiounbtqo0bN2r+/Pnat2+f+vTpU+w9Vq1apc8//1w9evRwB3eXLl00\nc+ZMZWdnu8O3pFH09X3j4uJu9lsFAKDS2SzDP9rscDjUs2dPdezY8ab25dqMZ+iX5+iZZ+iX5+iZ\nZ8rqV0V76YvtbnrEXNkmTJjgk30BAPAFI+9jBgDgdmV8MBcWFrqvXQMAUN1VOJiPHDmiESNG6K23\n3tJnn30mh8OhvXv3Vmjf48ePa86cOcWWjRkzpkL7btiwQbt27Spzm507d2r06NHas2dPhY4JAICp\nKnyNefXq1fLz81ODBg104sQJSdJf/vIXNWjQQI0aNdIf/vAHPffcc4qIiNDgwYO1cuVKBQUFKTQ0\nVH379pXT6dT8+fN16tQpJSUluT81PW/ePPn7++vIkSOaMWOGfvjhB33zzTfKz89XYmKitm3bpoKC\nAsXExCgwMFCS9Oqrr+rSpUs6d+6cpk2bppUrV6pu3bpq2bKlpGu/CEyePFm9evXSP/7xD3Xo0EHd\nunXzdu8AAPC6Cgdz37591bBhQ4WGhurMmTOSrt0q1b17d40fP14PP/yw2rRpo5deeklTpkxRcnKy\ngoKCNGHCBPXo0UPt27fX9OnTtWHDBm3YsEHStXuM+/btq0uXLum3336Ty+XS+fPnVaNGDT322GMK\nDQ3VAw88oJo1a7pD+fDhw7py5Ypmz56tAwcOaNWqVXrkkUcUGBio4OBgd72tWrXShAkTNH36dA0Y\nMMD9ywQAACbz6BqzZVnFHhxRu3ZtSVJgYKD8/f2LPXjC/QZ+frIsy73t9clJJOnkyZN6++23VbNm\nTbVt21aSFBUVpcTERB05ckRLly69YRYx6T/3KJe07ve1+fn5qWbNmp58mwAA+IxHt0vZbDb3H0la\nu3attm/frm7durnDVrp2/fill17SXXfdpS5dushutys7O1vvvPOOfv75Z82YMUMbNmxQrVq1VFRU\npMzMTB0/flyRkZHKycnRV199pTvuuEPR0dEKCgrS0qVL1bNnT9ntdrVu3Vr+/v5KSUlRXl6enn/+\neW3fvt27XQEAwEeMn2DEW7gx3zP0y3P0zDP0y3P0zDNVdYIR42+XAgDgdkIwAwBgEGODmYlFAAC3\no3KDmYlFAAC4dcr9VHZVm1hk4sSJiouL07/+9S/NmDFDzz//vJYvX15Z/QMAwKvKDeaqNrFIeHi4\nRo0apc2bN+vLL78s815nAABMU6FrzFVpYpGrV69Kki5duqQaNWq46wcAoCqoUDCXNLHIkiVLSp1Y\nJDU19YaJRXbs2KFHH31UkopNLLJv3z6dPXtWOTk5ev3113X06FHdf//9atq0qdatW6e8vDxJKjax\nyKpVq/T000+XGLh79+7V/PnztWPHDsXGxiomJkbTpk276UYBAHArVKsJRo4fP6533nlHSUlJN6zj\nxnzP0C/P0TPP0C/P0TPPMMGIAZo2bVpiKAMAUFVUq2AGAKCqI5gBADCIR0+XqgqOHDmiOXPmqFu3\nbmrevLlycnIUHR3t67IAAKiQajdiLm1CFAAAqoJqN2IubUIUAACqgmo3YpZKnxAFAADTVctgLmlC\nFAAAqoJqdyq7a9eu6tq1qyTp/vvvdy93Op2+KgkAgAqrdsEMAMB1ZQ3KKjpg8/Z25SGYAQDVVlWc\nwrRaXmMGAKCqIpgBADAIwQwAgEEIZgAADEIwAwBgEIIZAACDcLsUAOC2cfz4cS1ZskR2u13BwcF6\n9tlnS9xu9erVys7O1uXLl/Xdd98pMzOzxO1Onjwph8Mhu90uSZo+fXqJ2+3du1dpaWm6++671alT\nJz322GOl1siIGQBw20hLS1Pz5s2Vm5tb5j3Ow4cPV0pKiho1aqQlS5aUut2PP/6onTt36pdfflFw\ncHCp22VkZOjZZ5/V//3f/+mvf/2rCgoKSt2WYAYA3DZycnLUq1cvvfzyy3I4HGVu++OPP+rq1atq\n165dqds0atRIaWlpWrx4sXbv3l1q4I4cOVIrV67U/PnzdeHCBeXm5pZ6TIIZAHDbuPvuu1W3bl0F\nBAS4Tz+X5v3331diYmKZ26xevVrnz5+XJNntdl25cqXE7U6cOKHnnnvOfar7zjvvLPWYXGMGANw2\nRo8erYULFyooKEiPPvpomdsePnxYTZo0KXObQYMG6fXXX1fTpk0VERGhunXrlrhdw4YNNX/+fNWr\nV099+vRRQEDp8UswAwBuG61bt9Zrr71WoW3T0tLK3aZDhw5aunRpuds1a9ZMb775ZoXel1PZAAAY\nhGAGAMAgBDMAAAYhmAEAMAjBDACAQQhmAAAMQjADAGAQghkAAIMQzAAAGIRgBgDcdhwOh/bt2+fr\nMkrElJwAgGpr8uTJGjlypNavXy9/f3/VqlWr2DOTHQ6HevbsqY4dO2rMmDGaNm2a0tPTZbPZ1Lhx\nY40dO/aW18yIGQBQbcXFxSkkJET9+/dXZGSkdu3aVeb2y5YtU926dRUUFCSXy6WioqJbVOl/MGIG\nAFRbDz30kFJSUnTvvfcqMjJSgYGB7nWWZSkgIECXL1+WJOXm5sput2vo0KG65557tHbtWvn53frx\nK8EMAKi2/P391bhxY3333Xc6cOCACgsLZVmWJMlms+nBBx/Ua6+9ppYtW6pOnToaN26cUlNTFRIS\notDQUJ/UTDADAKq10aNH37BswoQJ7q+XL19ebJ3D4aj0msrCNWYAAAxCMAMAYBAjT2VblqWkpCQd\nPHhQgYGBmjt3rpo1a+Zen5GRoXfffVcBAQEKCwtTUlKS74oFAMCLjBwxb968WYWFhUpPT9fUqVOV\nkpLiXldQUKA33nhD7733nt5//32dP39eW7Zs8WG1AAB4j5HB7HQ6FR0dLUmKiIhQdna2e11gYKDS\n09PdH3m/cuWKatas6ZM6AQDwNiODOS8vT0FBQe7XAQEB7pu8bTabQkJCJEmrVq3SxYsX9cADD/ik\nTgAAvM3Ia8x2u135+fnu10VFRcVu8rYsS6+++qpycnI8+li70+n0ap3VHf3yHD3zDP3yHD2r/owM\n5sjISG3ZskX9+vWTy+VSWFhYsfWzZs1SrVq1tGTJEo+OGxUV5c0yqzWn00m/PETPPEO/PEfPPFNV\nf4kxMphjY2O1bds2JSQkSJJSUlKUkZGhixcvKjw8XOvXr1dUVJQSExNls9k0YsQIxcTE+LhqAABu\nnpHBbLPZlJycXGxZy5Yt3V/v37//VpcEAMAtYeSHvwAAuF0RzAAAGIRgBgDAIAQzAAAGIZgBADAI\nwQwAgEEIZgAADEIwAwBgEIIZAACDEMwAABiEYAYAwCAEMwAABiGYAQAwCMEMAIBBCGYAAAxCMAMA\nYBCCGQAAgxDMAAAYhGAGAMAgBDMAAAYhmAEAMAjBDACAQQhmAAAMQjADAGAQghkAAIMQzAAAGIRg\nBgDAIAQzAAAGIZgBADAIwQwAgEEIZgAADEIwAwBgEIIZAACDEMwAABiEYAYAwCAEMwAABiGYAQAw\nCMEMAIBBCGYAAAxCMAMAYBCCGQAAgxDMAAAYhGAGAMAgBDMAAAYhmAEAMAjBDACAQQhmAAAMQjAD\nAGAQghkAAIMQzAAAGIRgBgDAIAQzAAAGIZgBADAIwQwAgEEIZgAADEIwAwBgEIIZAACDEMwAABiE\nYAYAwCAEMwAABjEymC3L0pw5c5SQkKARI0bo6NGjxdZnZmZqyJAhSkhI0Nq1a31UJQAA3mdkMG/e\nvFmFhYVKT0/X1KlTlZKS4l535coVpaamasWKFVq1apU++OAD/fbbbz6sFgAA7zEymJ1Op6KjoyVJ\nERERys7Odq87fPiwWrRoIbvdrho1aigqKkpZWVm+KhUAAK8yMpjz8vIUFBTkfh0QEKCioqIS19Wt\nW1fnz5+/5TUCAFAZAnxdQEnsdrvy8/Pdr4uKiuTn5+del5eX516Xn5+vevXqVei4TqfTu4VWc/TL\nc/TMM/TLc/Ss+jMymCMjI7Vlyxb169dPLpdLYWFh7nWtW7dWTk6OcnNzVatWLWVlZWn06NHlHjMq\nKqoySwYAwCtslmVZvi7i9yzLUlJSkg4ePChJSklJ0d///nddvHhR8fHx2rp1qxwOhyzL0pAhQzR0\n6FAfVwygYBpFAAADxUlEQVQAgHcYGcwAANyujPzwFwAAtyuCGQAAgxDMAAAYhGAGAMAg1SqYmWPb\nc+X1LCMjQ0888YSGDRumpKQk3xRpkPL6dd3s2bO1aNGiW1ydmcrr2b59+zR8+HANHz5ckyZNUmFh\noY8qNUN5/fr00081aNAgxcfHa82aNT6q0jx79+5VYmLiDcur5M99qxr58ssvrRkzZliWZVkul8t6\n9tln3esuX75sxcbGWufPn7cKCwutwYMHW6dPn/ZVqcYoq2eXLl2yYmNjrYKCAsuyLGvKlClWZmam\nT+o0RVn9um7NmjXWk08+aS1cuPBWl2ek8nrWv39/66effrIsy7LWrl1rHTly5FaXaJTy+tWjRw8r\nNzfXKiwstGJjY63c3FxflGmUd955x4qLi7OefPLJYsur6s/9ajViZo5tz5XVs8DAQKWnpyswMFDS\ntQeI1KxZ0yd1mqKsfknSnj179P333yshIcEX5RmprJ4dOXJEwcHBSktLU2Jios6dO6fQ0FAfVWqG\n8v6PtWvXTufOnVNBQYEkyWaz3fIaTdOiRQu99dZbNyyvqj/3q1UwM8e258rqmc1mU0hIiCRp1apV\nunjxoh544AGf1GmKsvr166+/yuFwaPbs2bKYHsCtrJ6dOXNGLpdLiYmJSktL0/bt27Vr1y5flWqE\nsvolSW3bttXgwYP1+OOPq3fv3rLb7b4o0yixsbHy9/e/YXlV/blfrYK5subYrs7K6pl07XrX/Pnz\ntWPHDjkcDl+UaJSy+vXFF1/o7NmzeuaZZ7Rs2TJlZGTo448/9lWpxiirZ8HBwWrevLlatmypgIAA\nRUdH3zBCvN2U1a+DBw9q69atyszMVGZmpk6fPq1Nmzb5qlTjVdWf+9UqmCMjI/X1119LUplzbBcW\nFiorK0udOnXyVanGKKtnkjRr1ixdvnxZS5YscZ/Svp2V1a/ExEStW7dO7777rsaOHau4uDgNGDDA\nV6Uao6yeNWvWTBcuXHB/wMnpdKpNmzY+qdMUZfUrKChItWvXVmBgoPuMVm5urq9KNc7vz1RV1Z/7\nRj7E4n8VGxurbdu2ua/vpaSkKCMjwz3H9h//+EeNGjVKlmUpPj5eDRo08HHFvldWz8LDw7V+/XpF\nRUUpMTFRNptNI0aMUExMjI+r9p3y/o/hRuX1bO7cuZoyZYokqXPnzurVq5cvy/W58vp1/S6JwMBA\nNW/eXAMHDvRxxea4fr29qv/cZ65sAAAMUq1OZQMAUNURzAAAGIRgBgDAIAQzAAAGIZgBADAIwQwA\ngEEIZgAADPL/2Ij79/heonQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f0606634d10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(6.85, 6.1), dpi=300)\n",
    "\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax = fig.add_subplot(gs[0, 0])\n",
    "ax.text(0, 1, 'Figure 2',\n",
    "        size=16, va='top')\n",
    "ciepy.clean_axis(ax)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "gs.tight_layout(fig, rect=[0, 0.90, 0.5, 1])\n",
    "\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax = fig.add_subplot(gs[0, 0])\n",
    "t = dnase_res[dnase_res.biosample_type != 'immortalized cell line']\n",
    "t = dnase_res.ix[t.index].sort_values(by='pvalue', ascending=False).tail(10)\n",
    "# cdict = dict(zip(set(dnase_res.ix[t.index, 'biosample_type']), sns.color_palette('Set2')))\n",
    "cdict = dict([['induced pluripotent stem cell line', cpb.analysis.tableau20[0]],\n",
    "              ['stem cell', cpb.analysis.tableau20[4]],\n",
    "              ['primary cell', cpb.analysis.tableau20[8]]])\n",
    "#fig, ax = plt.subplots(1, 1, figsize=(5, 10))\n",
    "c = [cdict[x] for x in dnase_res.ix[t.index, 'biosample_type']]\n",
    "ax = (-np.log10(t.pvalue)).plot(kind='barh', ax=ax, color=c, label=None)\n",
    "ax.set_ylabel('')\n",
    "ax.set_xlabel('$-\\log_{10}$ $p$-value', fontsize=8)\n",
    "#ax.set_title('ENCODE DNase enrichments')\n",
    "ya, yb = ax.get_ylim()\n",
    "ax.set_yticklabels(dnase_res.ix[t.index, 'cell_type'], fontsize=8)\n",
    "rects = []\n",
    "labels = []\n",
    "for k in ['induced pluripotent stem cell line', 'stem cell',\n",
    "          'primary cell',]:\n",
    "    if k == 'induced pluripotent stem cell line':\n",
    "        labels.append('iPSC')\n",
    "    elif k == 'in vitro differentiated cells':\n",
    "        labels.append('in vitro\\ndifferentiated\\ncell')\n",
    "    else:\n",
    "        labels.append(k)\n",
    "    r = plt.Rectangle((0, 0), 0, 0, fc=cdict[k])\n",
    "    rects.append(r)\n",
    "lgd = ax.legend(rects, labels, loc='lower right', frameon=True, fontsize=7)\n",
    "#bbox_to_anchor=[0.5, -0.05])#, prop={'size':8}), loc='upper center',\n",
    "for p in lgd.get_patches():\n",
    "    p.set_linewidth(0)\n",
    "ax.grid(axis='y')\n",
    "for t in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "    t.set_fontsize(8)\n",
    "gs.tight_layout(fig, rect=[0, 0.27, 0.7, 0.59])\n",
    "\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax = fig.add_subplot(gs[0, 0])\n",
    "t = tf_res.sort_values(by='pvalue', ascending=False).tail(10)\n",
    "ax = (-np.log10(t.pvalue)).plot(kind='barh', ax=ax, color='grey', zorder=2)\n",
    "ax.set_ylabel('')\n",
    "ax.set_xlabel('$-\\log_{10}$ $p$-value', fontsize=8)\n",
    "ya, yb = ax.get_ylim()\n",
    "t = ax.set_yticklabels(t.target, fontsize=8)\n",
    "for t in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "    t.set_fontsize(8)\n",
    "ax.grid(axis='y')\n",
    "gs.tight_layout(fig, rect=[0.7, 0.27, 1, 0.59])\n",
    "\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax2 = fig.add_subplot(gs[0, 0])\n",
    "\n",
    "grasp_res.sort_values(by='pvalue', inplace=True)\n",
    "t = grasp_res[grasp_res.pvalue < 0.05]\n",
    "t = grasp_res[grasp_res.bh_sig]\n",
    "\n",
    "ymin = 0\n",
    "ymax= t.shape[0]\n",
    "\n",
    "ax2.set_xlabel('$-\\log_{10}$ $p$-value', fontsize=8)\n",
    "h = 0.5\n",
    "step = abs(ymax - ymin) / t.shape[0]\n",
    "b = np.arange(ymin, ymax, step) + (step / 2 - h / 2)\n",
    "ax2.barh(b, -np.log10(grasp_res.ix[t.index, 'pvalue']), height=h, \n",
    "         alpha=0.75, zorder=4)\n",
    "\n",
    "ax = ax2.twiny()\n",
    "\n",
    "ax = sns.boxplot(data=grasp_counts.ix[t.index, grasp_counts.columns[1:]].T, ax=ax, \n",
    "                 orient='h', linewidth=0.8)\n",
    "boxes = ax.artists\n",
    "for i,box in enumerate(boxes):\n",
    "    box.set_facecolor('grey')\n",
    "ax.scatter(grasp_counts.ix[t.index, 'real'], range(t.shape[0]), s=50, \n",
    "           color='black', marker='o', zorder=4)\n",
    "\n",
    "xmin, xmax = ax.get_xlim()\n",
    "# for i in range(se.shape[0]):\n",
    "#     ax.text(0.97 * xmax, i, se.values[i], va='center', ha='center')\n",
    "ax.set_xlabel('Number of overlaps', fontsize=8)\n",
    "ax.set_xlim(0, xmax)\n",
    "ax.set_ylabel('')\n",
    "for tick in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "    tick.set_fontsize(8)\n",
    "\n",
    "ymin,ymax = ax.get_ylim()\n",
    "\n",
    "ax.set_ylim(ymin, ymax)\n",
    "#ax.set_xlim(0, 50)\n",
    "# ax2.set_yticklabels(grasp_res.ix[t.index, 'phenotype'], fontsize=8)\n",
    "# for t in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "#     t.set_fontsize(8)\n",
    "\n",
    "ax.grid(axis='x')\n",
    "ax2.grid(axis='y')\n",
    "for tick in ax2.get_xticklabels() + ax2.get_yticklabels():\n",
    "    tick.set_fontsize(8)\n",
    "gs.tight_layout(fig, rect=[0, 0, 1, 0.31])\n",
    "\n",
    "# Number of eGenes vs. number of samples\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax = fig.add_subplot(gs[0, 0])\n",
    "ax.scatter(plotd.ix[plotd.source == 'gtex', 'num_samples'], \n",
    "           plotd.ix[plotd.source == 'gtex', 'num_sig_genes'],\n",
    "           label='GTEx', color=cpb.analysis.tableau20[0], s=30, alpha=0.75)\n",
    "ax.scatter([plotd.ix['ipsc_unrelateds', 'num_samples']], \n",
    "           [plotd.ix['ipsc_unrelateds', 'num_sig_genes']],\n",
    "           label='iPSC unrelateds', color=cpb.analysis.tableau20[2], alpha=0.75, s=30)\n",
    "ax.scatter([plotd.ix['ipsc', 'num_samples']], \n",
    "           [plotd.ix['ipsc', 'num_sig_genes']],\n",
    "           label='iPSC', color=cpb.analysis.tableau20[2], alpha=0.75, s=30, marker='*')\n",
    "ax.set_xlabel('Number of samples', fontsize=8)\n",
    "ax.set_ylabel('Number of eGenes', fontsize=8)\n",
    "for t in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "    t.set_fontsize(8)\n",
    "ax.legend(frameon=True, fancybox=True, fontsize=7, loc='lower right')\n",
    "ax.yaxis.set_major_formatter(ciepy.comma_format)\n",
    "gs.tight_layout(fig, rect=[0, 0.57, 0.5, 0.9])\n",
    "\n",
    "# Percent distinct eGenes\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax = fig.add_subplot(gs[0, 0])\n",
    "ax.scatter(plotd.ix[plotd.source == 'gtex', 'num_samples'], \n",
    "           plotd.ix[plotd.source == 'gtex', 'percent_distinct_egenes'],\n",
    "           label='GTEx', color=cpb.analysis.tableau20[0], s=30, alpha=0.75)\n",
    "ax.scatter([plotd.ix['ipsc_unrelateds', 'num_samples']], \n",
    "           [plotd.ix['ipsc_unrelateds', 'percent_distinct_egenes']],\n",
    "           label='iPSC 131 unrelateds', color=cpb.analysis.tableau20[2], alpha=0.75, s=30)\n",
    "ax.scatter([plotd.ix['ipsc', 'num_samples']], \n",
    "           [plotd.ix['ipsc', 'percent_distinct_egenes']],\n",
    "           label='iPSC', color=cpb.analysis.tableau20[2], alpha=0.75, s=50, marker='*')\n",
    "ax.set_xlabel('Number of samples', fontsize=8)\n",
    "ax.set_ylabel('Percent unique eGenes', fontsize=8)\n",
    "for t in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "    t.set_fontsize(8)\n",
    "#ax.legend(frameon=True, fancybox=True, fontsize=7, loc='upper right')\n",
    "ymin,ymax = ax.get_ylim()\n",
    "ax.set_ylim(0, ymax)\n",
    "gs.tight_layout(fig, rect=[0.5, 0.57, 1, 0.9])\n",
    "\n",
    "t = fig.text(0.005, 0.87, 'A', weight='bold', \n",
    "             size=12)\n",
    "t = fig.text(0.5, 0.87, 'B', weight='bold', \n",
    "             size=12)\n",
    "t = fig.text(0.005, 0.56, 'C', weight='bold', \n",
    "             size=12)\n",
    "t = fig.text(0.7, 0.56, 'D', weight='bold', \n",
    "             size=12)\n",
    "t = fig.text(0.005, 0.25, 'E', weight='bold', \n",
    "             size=12)\n",
    "\n",
    "fig.savefig(os.path.join(outdir, 'functional_annotation_power.pdf'))\n",
    "fig.savefig(os.path.join(outdir, 'functional_annotation_power.png'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Presentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(10, 6.5), dpi=300)\n",
    "\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax = fig.add_subplot(gs[0, 0])\n",
    "t = dnase_res[dnase_res.biosample_type != 'immortalized cell line']\n",
    "t = dnase_res.ix[t.index].sort_values(by='pvalue', ascending=False).tail(20)\n",
    "# cdict = dict(zip(set(dnase_res.ix[t.index, 'biosample_type']), sns.color_palette('Set2')))\n",
    "cdict = dict([['induced pluripotent stem cell line', cpb.analysis.tableau20[0]],\n",
    "              ['stem cell', cpb.analysis.tableau20[4]],\n",
    "              ['in vitro differentiated cells', cpb.analysis.tableau20[6]],\n",
    "              ['primary cell', cpb.analysis.tableau20[8]], \n",
    "              ['tissue', cpb.analysis.tableau20[10]]])\n",
    "#fig, ax = plt.subplots(1, 1, figsize=(5, 10))\n",
    "c = [cdict[x] for x in dnase_res.ix[t.index, 'biosample_type']]\n",
    "ax = (-np.log10(t.pvalue)).plot(kind='barh', ax=ax, color=c, label=None)\n",
    "ax.set_ylabel('')\n",
    "ax.set_xlabel('$-\\log_{10}$ $p$-value', fontsize=16)\n",
    "#ax.set_title('ENCODE DNase enrichments')\n",
    "ya, yb = ax.get_ylim()\n",
    "ax.set_yticklabels(dnase_res.ix[t.index, 'cell_type'], fontsize=8)\n",
    "rects = []\n",
    "labels = []\n",
    "for k in ['induced pluripotent stem cell line', 'stem cell',\n",
    "          'in vitro differentiated cells', 'primary cell', 'tissue']:\n",
    "    if k == 'induced pluripotent stem cell line':\n",
    "        labels.append('iPSC')\n",
    "    elif k == 'in vitro differentiated cells':\n",
    "        labels.append('in vitro\\ndifferentiated\\ncell')\n",
    "    else:\n",
    "        labels.append(k)\n",
    "    r = plt.Rectangle((0, 0), 0, 0, fc=cdict[k])\n",
    "    rects.append(r)\n",
    "lgd = ax.legend(rects, labels, loc='lower right', frameon=True, fontsize=12)\n",
    "#bbox_to_anchor=[0.5, -0.05])#, prop={'size':8}), loc='upper center',\n",
    "for p in lgd.get_patches():\n",
    "    p.set_linewidth(0)\n",
    "ax.grid(axis='y')\n",
    "for t in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "    t.set_fontsize(12)\n",
    "ax.set_title('ENCODE DHS')\n",
    "gs.tight_layout(fig, rect=[0, 0, 0.6, 1])\n",
    "\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax = fig.add_subplot(gs[0, 0])\n",
    "t = tf_res.sort_values(by='pvalue', ascending=False).tail(20)\n",
    "ax = (-np.log10(t.pvalue)).plot(kind='barh', ax=ax, color='grey')\n",
    "ax.set_ylabel('')\n",
    "ax.set_xlabel('$-\\log_{10}$ $p$-value', fontsize=16)\n",
    "ya, yb = ax.get_ylim()\n",
    "t = ax.set_yticklabels(t.target, fontsize=8)\n",
    "for t in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "    t.set_fontsize(12)\n",
    "ax.grid(axis='y')\n",
    "ax.set_title('H1-hESC Transcription Factor ChIP-Seq')\n",
    "gs.tight_layout(fig, rect=[0.6, 0, 1, 1])\n",
    "\n",
    "fig.savefig(os.path.join(outdir, 'dnase_tf_presentation.pdf'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(1, 1, figsize=(10, 6.5), dpi=300)\n",
    "\n",
    "grasp_res.sort_values(by='pvalue', inplace=True)\n",
    "t = grasp_res[grasp_res.bh_sig]\n",
    "\n",
    "sns.boxplot(data=grasp_counts.ix[t.index, grasp_counts.columns[1:]].T, ax=ax, orient='h')\n",
    "ax.scatter(grasp_counts.ix[t.index, 'real'], range(t.shape[0]), alpha=0.5, s=100)\n",
    "\n",
    "xmin, xmax = ax.get_xlim()\n",
    "# for i in range(se.shape[0]):\n",
    "#     ax.text(0.97 * xmax, i, se.values[i], va='center', ha='center')\n",
    "ax.set_xlabel('Number of overlaps', fontsize=16)\n",
    "ax.set_xlim(0, xmax)\n",
    "ax.set_ylabel('')\n",
    "for tick in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "    tick.set_fontsize(16)\n",
    "\n",
    "ymin,ymax = ax.get_ylim()\n",
    "ax2 = ax.twiny()\n",
    "ax2.set_xlabel('$-\\log_{10}$ Fisher $p$-value', fontsize=16)\n",
    "h = 0.5\n",
    "step = abs(ymax - ymin) / t.shape[0]\n",
    "b = np.arange(ymax, ymin, step) + (step / 2 - h / 2)\n",
    "ax2.barh(b, -np.log10(grasp_res.ix[t.index, 'pvalue']), height=h, alpha=0.5)\n",
    "ax.set_ylim(ymin, ymax)\n",
    "#ax.set_xlim(0, 50)\n",
    "ax.grid(axis='x')\n",
    "ax2.grid(axis='x')\n",
    "ax.grid('x', zorder=4)\n",
    "for tick in ax2.get_xticklabels() + ax2.get_yticklabels():\n",
    "    tick.set_fontsize(16)\n",
    "    \n",
    "fig.tight_layout()\n",
    "\n",
    "fig.savefig(os.path.join(outdir, 'gwas_presentation.pdf'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fs = 10\n",
    "\n",
    "fig = plt.figure(figsize=(6.8, 4), dpi=300)\n",
    "\n",
    "# Number of eGenes vs. number of samples\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax = fig.add_subplot(gs[0, 0])\n",
    "ax.scatter(plotd.ix[plotd.source == 'gtex', 'num_samples'], \n",
    "           plotd.ix[plotd.source == 'gtex', 'num_sig_genes'],\n",
    "           label='GTEx', color=cpb.analysis.tableau20[0], s=30, alpha=0.75)\n",
    "ax.scatter([plotd.ix['ipsc_unrelateds', 'num_samples']], \n",
    "           [plotd.ix['ipsc_unrelateds', 'num_sig_genes']],\n",
    "           label='iPSC 131 unrelateds', color=cpb.analysis.tableau20[2], alpha=0.75, s=50,\n",
    "           marker='*')\n",
    "ax.scatter([plotd.ix['ipsc', 'num_samples']], \n",
    "           [plotd.ix['ipsc', 'num_sig_genes']],\n",
    "           label='iPSC', color=cpb.analysis.tableau20[2], alpha=0.75, s=30)\n",
    "ax.set_xlabel('Number of samples', fontsize=fs)\n",
    "ax.set_ylabel('Number of eGenes', fontsize=fs)\n",
    "for t in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "    t.set_fontsize(fs)\n",
    "lgd = ax.legend(loc='upper center', bbox_to_anchor=(1.35, 1))\n",
    "ax.yaxis.set_major_formatter(ciepy.comma_format)\n",
    "gs.tight_layout(fig, rect=[0, 0.5, 0.5, 1])\n",
    "\n",
    "# Percent distinct eGenes\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax = fig.add_subplot(gs[0, 0])\n",
    "ax.scatter(plotd.ix[plotd.source == 'gtex', 'num_samples'], \n",
    "           plotd.ix[plotd.source == 'gtex', 'percent_distinct_egenes'],\n",
    "           label='GTEx', color=cpb.analysis.tableau20[0], s=30, alpha=0.75)\n",
    "ax.scatter([plotd.ix['ipsc_unrelateds', 'num_samples']], \n",
    "           [plotd.ix['ipsc_unrelateds', 'percent_distinct_egenes']],\n",
    "           label='iPSC 131 unrelateds', color=cpb.analysis.tableau20[2], alpha=0.75, s=30)\n",
    "ax.scatter([plotd.ix['ipsc', 'num_samples']], \n",
    "           [plotd.ix['ipsc', 'percent_distinct_egenes']],\n",
    "           label='iPSC', color=cpb.analysis.tableau20[2], alpha=0.75, s=50, marker='*')\n",
    "ax.set_xlabel('Number of samples', fontsize=fs)\n",
    "ax.set_ylabel('Percent unique eGenes', fontsize=fs)\n",
    "for t in ax.get_xticklabels() + ax.get_yticklabels():\n",
    "    t.set_fontsize(fs)\n",
    "#ax.legend(frameon=True, fancybox=True, fontsize=fs, loc='upper right')\n",
    "ymin,ymax = ax.get_ylim()\n",
    "ax.set_ylim(0, ymax)\n",
    "gs.tight_layout(fig, rect=[0, 0, 0.5, 0.5])\n",
    "\n",
    "fig.savefig(os.path.join(outdir, 'power_presentation.pdf'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Environment (cie)",
   "language": "",
   "name": "cie"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
