{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imprinting\n",
    "\n",
    "I'd like to look at imprinting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import cPickle\n",
    "import datetime\n",
    "import glob\n",
    "import os\n",
    "import random\n",
    "import re\n",
    "import subprocess\n",
    "\n",
    "import cdpybio as cpb\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "import pybedtools as pbt\n",
    "import scipy.stats as stats\n",
    "import seaborn as sns\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import vcf as pyvcf\n",
    "\n",
    "import cardipspy as cpy\n",
    "import ciepy\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "import socket\n",
    "if socket.gethostname() == 'fl-hn1' or socket.gethostname() == 'fl-hn2':\n",
    "    pbt.set_tempdir('/frazer01/home/cdeboever/tmp')\n",
    "    \n",
    "outdir = os.path.join(ciepy.root, 'output',\n",
    "                      'imprinting')\n",
    "cpy.makedir(outdir)\n",
    "\n",
    "private_outdir = os.path.join(ciepy.root, 'private_output',\n",
    "                              'imprinting')\n",
    "cpy.makedir(private_outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fn = os.path.join(ciepy.root, 'output', 'input_data', 'rsem_tpm.tsv')\n",
    "tpm = pd.read_table(fn, index_col=0)\n",
    "fn = os.path.join(ciepy.root, 'output', 'input_data', 'rnaseq_metadata.tsv')\n",
    "rna_meta = pd.read_table(fn, index_col=0)\n",
    "fn = os.path.join(ciepy.root, 'output', 'input_data', 'subject_metadata.tsv')\n",
    "subject_meta = pd.read_table(fn, index_col=0)\n",
    "fn = os.path.join(ciepy.root, 'output', 'input_data', 'wgs_metadata.tsv')\n",
    "wgs_meta = pd.read_table(fn, index_col=0)\n",
    "\n",
    "gene_info = pd.read_table(cpy.gencode_gene_info, index_col=0)\n",
    "gene_info['ensembl_id'] = [x.split('.')[0] for x in gene_info.index]\n",
    "\n",
    "genes = pbt.BedTool(cpy.gencode_gene_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fn = os.path.join(ciepy.root, 'output', 'input_data', \n",
    "                  'mbased_major_allele_freq.tsv')\n",
    "maj_af = pd.read_table(fn, index_col=0)\n",
    "\n",
    "fn = os.path.join(ciepy.root, 'output', 'input_data', \n",
    "                  'mbased_p_val_ase.tsv')\n",
    "ase_pval = pd.read_table(fn, index_col=0)\n",
    "\n",
    "locus_p = pd.Panel({'major_allele_freq':maj_af, 'p_val_ase':ase_pval})\n",
    "locus_p = locus_p.swapaxes(0, 2)\n",
    "\n",
    "snv_fns = glob.glob(os.path.join(ciepy.root, 'private_output', 'input_data', 'mbased_snv',\n",
    "                                 '*_snv.tsv'))\n",
    "count_fns = glob.glob(os.path.join(ciepy.root, 'private_output', 'input_data', 'allele_counts',\n",
    "                                   '*mbased_input.tsv'))\n",
    "\n",
    "snv_res = {}\n",
    "for fn in snv_fns:\n",
    "    snv_res[os.path.split(fn)[1].split('_')[0]] = pd.read_table(fn, index_col=0)\n",
    "    \n",
    "count_res = {}\n",
    "for fn in count_fns:\n",
    "    count_res[os.path.split(fn)[1].split('_')[0]] = pd.read_table(fn, index_col=0)\n",
    "\n",
    "snv_p = pd.Panel(snv_res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imprinted Genes\n",
    "\n",
    "I'll get some known imprinted genes.\n",
    "\n",
    "### geneimprint.com\n",
    "\n",
    "I'm going to get the imprinted genes from geneimprint.com. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fn = os.path.join(outdir, 'geneimprint.tsv')\n",
    "if not os.path.exists(fn):\n",
    "    geneimprint = pd.read_html('http://www.geneimprint.com/site/genes-by-species.Homo+sapiens')[0]\n",
    "    geneimprint.columns = geneimprint.ix[0]\n",
    "    geneimprint = geneimprint.drop(0)\n",
    "    geneimprint = geneimprint[geneimprint.Status.apply(lambda x: x in ['Predicted', 'Imprinted'])]\n",
    "    \n",
    "    # Now I'll convert the geneimprint gene names into gencode gene IDs. I'll drop any genes I can't\n",
    "    # convert.\n",
    "    ind = []\n",
    "    drop = []\n",
    "    for i in geneimprint.index:\n",
    "        name = geneimprint.ix[i, 'Gene'].strip('*')\n",
    "        tdf = gene_info[gene_info.gene_name == name]\n",
    "        if tdf.shape[0] == 1:\n",
    "            ind.append(tdf.index[0])\n",
    "        else:\n",
    "            names = []\n",
    "            if type(geneimprint.ix[i, 'Aliases']) == unicode:\n",
    "                names += [x.strip() for x in geneimprint.ix[i, 'Aliases'].split(',')]\n",
    "            tdf = gene_info[gene_info.gene_name.apply(lambda x: x in names)].drop_duplicates()\n",
    "            if tdf.shape[0] == 1:\n",
    "                ind.append(tdf.index[0])\n",
    "            else:\n",
    "                drop.append(i)\n",
    "    print('Dropping {} geneimprint genes'.format(len(drop)))\n",
    "    geneimprint = geneimprint.drop(drop)\n",
    "    geneimprint.index = ind\n",
    "    for c in geneimprint.columns:\n",
    "        ind = geneimprint[geneimprint[c].isnull() == False].index\n",
    "        geneimprint.ix[ind, c] = geneimprint.ix[ind, c].apply(lambda x: x.encode('ascii', 'ignore'))\n",
    "    geneimprint.columns = [x.encode('ascii', 'ignore') for x in geneimprint.columns]\n",
    "    geneimprint.to_csv(fn, sep='\\t')\n",
    "else:\n",
    "    geneimprint = pd.read_table(fn, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [Babak et al. 2015](http://www.ncbi.nlm.nih.gov/pubmed/258487520)\n",
    "\n",
    "Human imprinted genes from Babak. The values in `babak_cs` are their \"combined scores.\"\n",
    "A combined score greater than zero is evidence for imprinting. `babak_cs` has the combined score\n",
    "for 94 genes in 45 different different tissues. \n",
    "\n",
    "I'm not going to filter the genes from Babak yet. I may want to remove genes from other\n",
    "sources if they don't have imprinting in many tissues according to Babak.\n",
    "\n",
    "TODO: Can I get imprinting direction from Babak?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fn = os.path.join(outdir, 'babak.tsv')\n",
    "if not os.path.exists(fn):\n",
    "    babak_cs = pd.read_excel('http://www.nature.com/ng/journal/v47/n5/extref/ng.3274-S11.xlsx',\n",
    "                             sheetname='SuppFile_5', skiprows=213, skip_footer=503-309, index_col=0)\n",
    "    babak_cs = babak_cs > 0\n",
    "\n",
    "    # Now I'll convert the Babak gene names into gencode gene IDs. I'll drop any genes I can't\n",
    "    # convert.\n",
    "    ind = []\n",
    "    drop = []\n",
    "    for i in babak_cs.index:\n",
    "        tdf = gene_info[gene_info.gene_name == i.strip('-new')]\n",
    "        if tdf.shape[0] == 1:\n",
    "            ind.append(tdf.index[0])\n",
    "        else:\n",
    "            drop.append(i)\n",
    "    print('Dropping {} Babak genes'.format(len(drop)))\n",
    "    babak_cs = babak_cs.drop(drop)\n",
    "    babak_cs.index = ind\n",
    "    babak_cs.to_csv(fn, sep='\\t')\n",
    "else:\n",
    "    babak_cs = pd.read_table(fn, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [Baran et al. 2015](http://genome.cshlp.org/content/25/7/927.full)\n",
    "\n",
    "I'm not going to filter the genes from Babak yet. I may want to remove genes from other\n",
    "sources if they don't have imprinting in many tissues according to Baran."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fn = os.path.join(outdir, 'baran.tsv')\n",
    "if not os.path.exists(fn):\n",
    "    url = 'http://genome.cshlp.org/content/suppl/2015/05/20/gr.192278.115.DC1/Supp_Tables_S1-4_6-7.xlsx'\n",
    "    baran = pd.read_excel(url, sheetname='Table_S6')\n",
    "    baran_fam = pd.read_excel(url, sheetname='Table_S7')\n",
    "\n",
    "    # I want to figure out which genes validated in the family data and what their imprinting status is .\n",
    "    g = set(baran_fam.Gene) & set(baran.Human_Genename)\n",
    "    baran_status = {}\n",
    "    for gene in g:\n",
    "        t = baran_fam[baran_fam.Gene == gene]\n",
    "        if sum(t.Imprinting_status.apply(lambda x: 'Paternally expressed' in x)) == t.shape[0]:\n",
    "            baran_status[gene] = 'Paternal'\n",
    "        elif sum(t.Imprinting_status.apply(lambda x: 'Maternally expressed' in x)) == t.shape[0]:\n",
    "            baran_status[gene] = 'Maternal'\n",
    "    baran['ExpressedAllele'] = np.nan\n",
    "    # Note that I'm requiring all tissues to have the same direction to consider something validated.\n",
    "    # This is stringent.\n",
    "    baran['validated'] = np.nan\n",
    "    for gene in baran_status.keys():\n",
    "        baran.ix[baran.Human_Genename == gene, 'ExpressedAllele'] = baran_status[gene]\n",
    "        baran.ix[baran.Human_Genename == gene, 'validated'] = True\n",
    "    for gene in set(baran_fam.Gene) - set(baran_status.keys()):\n",
    "        baran.ix[baran.Human_Genename == gene, 'validated'] = False\n",
    "    se = pd.Series(gene_info.index, index=gene_info.ensembl_id)\n",
    "    baran['gencode_id'] = se[baran.Human_GeneID].values\n",
    "    baran = baran.dropna(subset=['gencode_id'])\n",
    "    baran.index = baran.gencode_id\n",
    "    baran.to_csv(fn, sep='\\t')\n",
    "else:\n",
    "    baran = pd.read_table(fn, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine different sources\n",
    "\n",
    "I'll keep the genes that have evidence for imprinting in more than 5 tissues \n",
    "from Babak et al. and that were marked as \"imprinted\" or \"consistent with imprinted\"\n",
    "from Baran et al."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fn = os.path.join(outdir, 'igenes.tsv')\n",
    "if not os.path.exists(fn):\n",
    "    babak_cs_f = babak_cs[babak_cs.sum(axis=1) >= 5]\n",
    "    babak_not_imp = set(babak_cs.index) - set(babak_cs_f.index)\n",
    "\n",
    "    vals = ['imprinted', 'consistent with imprinted']\n",
    "    baran_f = baran[(baran.Imprinting_status_Baran.apply(lambda x: x in vals)) & \n",
    "                    (baran.validated != False)]\n",
    "    baran_not_imp = set(baran.index) - set(baran_f.index)\n",
    "\n",
    "    # Now I'll make a dataframe indexed by Gencode gene ID and filled in with some\n",
    "    # of the info I have available.\n",
    "    g = set(geneimprint.index) | set(babak_cs_f.index) | set(baran_f.index)\n",
    "    igenes = pd.DataFrame(index=g, columns=['geneimprint', 'babak', 'baran', \n",
    "                                            'expressed_allele', 'num_babak_tissues'])\n",
    "\n",
    "    igenes['geneimprint'] = False\n",
    "    igenes.ix[geneimprint.index, 'geneimprint'] = True\n",
    "    igenes['babak'] = False\n",
    "    igenes.ix[babak_cs_f.index, 'babak'] = True\n",
    "    igenes['baran'] = False\n",
    "    igenes.ix[baran_f.index, 'baran'] = True\n",
    "\n",
    "    # I thought about dropping genes with evidence against imprinting from Babak or Baran\n",
    "    # but I think I want to be more inclusive. I think the papers could have false negatives.\n",
    "    # I see a stronger enrichment for ASE when I don't remove these genes (arguing some are \n",
    "    # printed) although the enrichment exists either way.\n",
    "    # igenes = igenes.drop((babak_not_imp | baran_not_imp) & set(igenes.index))\n",
    "\n",
    "    ind = set(geneimprint.index) & set(igenes.index)\n",
    "    igenes.ix[ind, 'expressed_allele'] = geneimprint.ix[ind, 'ExpressedAllele']\n",
    "\n",
    "    # Add info on imprinting status if I have it.\n",
    "    t = baran_f.dropna(subset=['ExpressedAllele'])\n",
    "    ind = set(t.index) & set(igenes.index)\n",
    "    igenes.ix[ind, 'expressed_allele'] = t.ix[ind, 'ExpressedAllele']\n",
    "\n",
    "    # Add info on number of tissues imprinted in.\n",
    "    se = babak_cs.sum(axis=1)\n",
    "    ind = set(igenes.index) & set(se.index)\n",
    "    igenes.ix[ind, 'num_babak_tissues'] = se[ind].values\n",
    "    igenes.to_csv(fn, sep='\\t')\n",
    "else:\n",
    "    igenes = pd.read_table(fn, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO: I may want to try to convert more of the genes from geneimprint. I can use mygene.info:\n",
    "\n",
    "    import mygene\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    res = mg.query('symbol:HYMAI', species='human')\n",
    "\n",
    "## ASE\n",
    "\n",
    "Are imprinted genes enriched for having ASE?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "t = locus_p.ix[:, :, 'p_val_ase']\n",
    "t_igenes = t.ix[set(t.index) & set(igenes.index) - set(gene_info[gene_info.chrom == 'chrX'].index)]\n",
    "t_igenes = t_igenes[t_igenes.isnull().sum(axis=1) / t_igenes.shape[1] < 0.5]\n",
    "t_not_igenes = t.ix[set(t.index) - set(igenes.index) - set(gene_info[gene_info.chrom == 'chrX'].index)]\n",
    "t_not_igenes = t_not_igenes[t_not_igenes.isnull().sum(axis=1) / t_not_igenes.shape[1] < 0.5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "a = (t_igenes < 0.005).sum() / (t_igenes.isnull() == False).sum()\n",
    "b = (t_not_igenes < 0.005).sum() / (t_not_igenes.isnull() == False).sum()\n",
    "print('Average number of significant ASE samples per gene ' \n",
    "      'for {} imprinted genes: {:.2f}%.'.format(t_igenes.shape[0], a.mean() * 100))\n",
    "print('Average number of significant ASE samples per gene ' \n",
    "      'for {} non-imprinted genes: {:.2f}%.'.format(t_not_igenes.shape[0], b.mean() * 100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "r = stats.mannwhitneyu(a, b)\n",
    "print('The average number of samples with ASE is significantly '\n",
    "      'different between imprinted and non-imprinted genes with p={:.3e}.'.format(r.pvalue))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "log_tpm = np.log10(tpm + tpm[tpm > 0].min().min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "r = stats.mannwhitneyu(log_tpm.ix[t_not_igenes.index].mean(axis=1), log_tpm.ix[t_igenes.index].mean(axis=1))\n",
    "print('The average TPM expression between the imprinted and non-imprinted genes is comparable '\n",
    "      'p={:.2f}.'.format(r.pvalue))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CARDiPS Trios and Phasing\n",
    "\n",
    "I can find CARDiPS subjects for which we have WGS for both parents so that I can \n",
    "accurately assign parental haplotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24 trio children\n"
     ]
    }
   ],
   "source": [
    "# I'll keep children where we have WGS data for both parents and RNA-seq data for the child.\n",
    "trio_child = subject_meta.dropna(subset=['father_id', 'mother_id'])\n",
    "trio_child = trio_child[trio_child.father_id.apply(lambda x: x in wgs_meta.subject_id.values)]\n",
    "trio_child = trio_child[trio_child.mother_id.apply(lambda x: x in wgs_meta.subject_id.values)]\n",
    "trio_child = trio_child[[x in rna_meta.subject_id.values for x in trio_child.index]]\n",
    "print('{} trio children'.format(trio_child.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's filter the imprinted genes to those where we estimated ASE in at \n",
    "least one trio child."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trio_child['wgs_id'] = [wgs_meta[wgs_meta.subject_id == x].index[0] for x in trio_child.index]\n",
    "trio_child['father_wgs_id'] = [wgs_meta[wgs_meta.subject_id == x].index[0] for x in trio_child.father_id]\n",
    "trio_child['mother_wgs_id'] = [wgs_meta[wgs_meta.subject_id == x].index[0] for x in trio_child.mother_id]\n",
    "def get_rna_sample(sid):\n",
    "    tdf = rna_meta[rna_meta.subject_id == sid]\n",
    "    if tdf.shape[0] == 1:\n",
    "        rna_sample = tdf.index[0]\n",
    "    elif tdf[tdf.in_eqtl].shape[0] == 1:\n",
    "        rna_sample = tdf[tdf.in_eqtl].index[0]\n",
    "    elif tdf[tdf.in_222].shape[0] == 1:\n",
    "        rna_sample = tdf[tdf.in_22].index[0]\n",
    "    else:\n",
    "        rna_sample = np.nan\n",
    "    return rna_sample\n",
    "trio_child['rna_id'] = [get_rna_sample(x) for x in trio_child.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trio_child_ase_pval = ase_pval[trio_child.rna_id]\n",
    "igenes_f = igenes.ix[set(igenes.index) & set(trio_child_ase_pval.index)]\n",
    "trio_child_ase_pval = trio_child_ase_pval.ix[igenes_f.index]\n",
    "trio_child_ase_pval = trio_child_ase_pval.ix[set(trio_child_ase_pval.index) - \n",
    "                                             set(gene_info[gene_info.chrom == 'chrX'].index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def imprinting_analysis(child, gene_id):\n",
    "    # Make data frame with SNV-level results for this sample and gene.\n",
    "    snv_df = snv_p.ix[trio_child.ix[child, 'rna_id']]\n",
    "    snv_df = snv_df[snv_df.locus == gene_id]\n",
    "\n",
    "    # Let's get a 1MB region centered on the gene and get all variants in this\n",
    "    # region. We'll look for variants that are het in the child and homozygous\n",
    "    # alt/ref in the parents so we can determine the parental haplotype.\n",
    "    chrom = gene_info.ix[gene_id, 'chrom']\n",
    "    start = gene_info.ix[gene_id, 'start']\n",
    "    end = gene_info.ix[gene_id, 'end']\n",
    "    slop = (1000000 - (end - start)) / 2\n",
    "    bt = pbt.BedTool('{}\\t{}\\t{}'.format(chrom, start, end), from_string=True)\n",
    "    bt = bt.slop(b=slop, g=pbt.genome_registry.hg19)\n",
    "    df = bt.to_dataframe()\n",
    "    if df.ix[0, 'end'] - df.ix[0, 'start'] != 1000000:\n",
    "        if df.ix[0, 'start'] == 0:\n",
    "            bt = bt.slop(l=0, r=1000000 - (df.ix[0, 'end'] - df.ix[0, 'start']), \n",
    "                         g=pbt.genome_registry.hg19)\n",
    "            df = bt.to_dataframe()\n",
    "        else:\n",
    "            bt = bt.slop(r=0, l=1000000 - (df.ix[0, 'end'] - df.ix[0, 'start']), \n",
    "                         g=pbt.genome_registry.hg19)\n",
    "            df = bt.to_dataframe()\n",
    "    start = df.ix[0, 'start']\n",
    "    end = df.ix[0, 'end']\n",
    "    os.remove(bt.fn)\n",
    "\n",
    "    fn = ('/projects/CARDIPS/pipeline/WGS/mergedVCF/phased_20151214/'\n",
    "          'CARDIPS_{}_phased.vcf.gz'.format(gene_info.ix[gene_id, 'chrom']))\n",
    "    vcf_reader = pyvcf.Reader(open(fn))\n",
    "    res = vcf_reader.fetch(gene_info.ix[gene_id, 'chrom'][3:], start, end)\n",
    "\n",
    "    variants = []\n",
    "    for r in res:\n",
    "        g = r.genotype(trio_child.ix[child, 'wgs_id'])\n",
    "        if g.called and g.is_het:\n",
    "            mg = r.genotype(trio_child.ix[child, 'mother_wgs_id'])\n",
    "            fg = r.genotype(trio_child.ix[child, 'father_wgs_id'])\n",
    "            if mg.called and fg.called:\n",
    "                if set(mg.gt_alleles) == set('0') and set(fg.gt_alleles) == set('1'):\n",
    "                    variants.append(r)\n",
    "                elif set(mg.gt_alleles) == set('1') and set(fg.gt_alleles) == set('0'):\n",
    "                    variants.append(r)\n",
    "\n",
    "    left_hap_father = 0\n",
    "    for r in variants:\n",
    "        g = r.genotype(trio_child.ix[child, 'wgs_id'])\n",
    "        fg = r.genotype(trio_child.ix[child, 'father_wgs_id'])\n",
    "        if g.gt_alleles[0] == fg.gt_alleles[0]:\n",
    "            left_hap_father += 1\n",
    "\n",
    "    if left_hap_father > len(variants) / 2:\n",
    "        father_hap = 0\n",
    "    else:\n",
    "        father_hap = 1\n",
    "\n",
    "    snv_df.loc[:, 'preference'] = ''\n",
    "    for i in snv_df.index:\n",
    "        r = vcf_reader.fetch(gene_info.ix[gene_id, 'chrom'][3:], snv_df.ix[i, 'position'])\n",
    "        g = r.genotype(trio_child.ix[child, 'wgs_id'])\n",
    "        if snv_df.ix[i, 'ref_is_major'] and g.gt_alleles[father_hap] == '0':\n",
    "            snv_df.loc[i, 'preference'] = 'paternal'\n",
    "        elif snv_df.ix[i, 'ref_is_major'] and g.gt_alleles[father_hap] == '1':\n",
    "            snv_df.loc[i, 'preference'] = 'maternal'\n",
    "        elif snv_df.ix[i, 'ref_is_major'] == False and g.gt_alleles[father_hap] == '0':\n",
    "            snv_df.loc[i, 'preference'] = 'maternal'\n",
    "        elif snv_df.ix[i, 'ref_is_major'] == False and g.gt_alleles[father_hap] == '1':\n",
    "            snv_df.loc[i, 'preference'] = 'paternal'\n",
    "        else:\n",
    "            print('problem for {}'.format(i))\n",
    "#     vc = snv_df.preference.value_counts()\n",
    "#     if vc.shape[0] == 1:\n",
    "#         preference = vc.index[0]\n",
    "#     if vc.shape[0] == 2:\n",
    "#         if vc['maternal'] == vc['paternal']:\n",
    "#             preference = 'unknown'\n",
    "#         else:\n",
    "#             prefernce = vc.index[0]\n",
    "    return (gene_id, child, ase_pval.ix[gene_id, trio_child.ix[child, 'rna_id']], snv_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fn = os.path.join(outdir, 'results.pickle')\n",
    "if not os.path.exists(fn):\n",
    "    from ipyparallel import Client\n",
    "    #parallel_client = Client(profile='parallel')\n",
    "    parallel_client = Client()\n",
    "    dview = parallel_client[:]\n",
    "    print('Cluster has {} engines.'.format(len(parallel_client.ids)))\n",
    "    \n",
    "    with dview.sync_imports():\n",
    "        import cdpybio\n",
    "        import cardipspy\n",
    "        import pybedtools\n",
    "        import os\n",
    "        import vcf\n",
    "\n",
    "    %px cpb = cdpybio\n",
    "    %px cpy = cardipspy\n",
    "    %px pbt = pybedtools\n",
    "    %px pyvcf = vcf\n",
    "\n",
    "    dview.push(dict(imprinting_analysis=imprinting_analysis, snv_p=snv_p, \n",
    "                    gene_info=gene_info, trio_child=trio_child, ase_pval=ase_pval));\n",
    "\n",
    "    pairs = []\n",
    "    for gene_id in trio_child_ase_pval.index:\n",
    "        se = trio_child_ase_pval.ix[gene_id].dropna()\n",
    "        for rna_id in se.index:\n",
    "            child = trio_child[trio_child.rna_id == rna_id].index[0]\n",
    "            pairs.append((child, gene_id))\n",
    "\n",
    "    def chunks(l, n):\n",
    "        \"\"\"Yield successive n-sized chunks from l.\"\"\"\n",
    "        for i in xrange(0, len(l), n):\n",
    "            yield l[i:i+n]\n",
    "\n",
    "    # It would be faster if I didn't do this in chunks but I ran into an error\n",
    "    # and this fixed it. However, I think I should probably try them all at the \n",
    "    # same time again at some point.\n",
    "    c = chunks(pairs, len(dview))\n",
    "    results = []\n",
    "    while True:\n",
    "        try:\n",
    "            results += dview.map_sync(lambda x: imprinting_analysis(x[0], x[1]), c.next())\n",
    "        except StopIteration:\n",
    "            break\n",
    "            \n",
    "#     results = dview.map_sync(lambda x: imprinting_analysis(x[0], x[1]), pairs[0:10])\n",
    "#     results = results.get()\n",
    "            \n",
    "    with open(fn , 'w') as f:\n",
    "        cPickle.dump(results, f)\n",
    "\n",
    "else:\n",
    "    results = cPickle.load(open(fn))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.DataFrame([x[0:3] for x in results], columns=['gene_id', 'sample_id', 'p_value'])\n",
    "df['maternal_percent'] = 0\n",
    "df['paternal_percent'] = 0\n",
    "for i,x in enumerate(results):\n",
    "    vc = x[-1].preference.value_counts()\n",
    "    if 'paternal' in vc.index:\n",
    "        df.ix[i, 'paternal_percent'] = vc['paternal'] / float(vc.sum())\n",
    "    if 'maternal' in vc.index:\n",
    "        df.ix[i, 'maternal_percent'] = vc['maternal'] / float(vc.sum())\n",
    "df['num_snvs'] = [x[-1].shape[0] for x in results]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = df.merge(igenes_f.dropna(subset=['expressed_allele'])[['expressed_allele']], \n",
    "              left_on='gene_id', right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imprinted genes are more likely to show bias in the direction they are imprinted: odds=1.680, p=7.21e-04.\n"
     ]
    }
   ],
   "source": [
    "pat = df.ix[df.paternal_percent > df.maternal_percent, 'gene_id']\n",
    "mat = df.ix[df.maternal_percent > df.paternal_percent, 'gene_id']\n",
    "pat_pat = sum(igenes_f.ix[pat, 'expressed_allele'] == 'Paternal')\n",
    "mat_mat = sum(igenes_f.ix[mat, 'expressed_allele'] == 'Maternal')\n",
    "pat_mat = sum(igenes_f.ix[pat, 'expressed_allele'] == 'Maternal')\n",
    "mat_pat = sum(igenes_f.ix[mat, 'expressed_allele'] == 'Paternal')\n",
    "\n",
    "odds, pvalue = stats.fisher_exact([[pat_pat, pat_mat], [mat_pat, mat_mat]])\n",
    "print('Imprinted genes are more likely to show bias in '\n",
    "      'the direction they are imprinted: odds={:.3f}, p={:.2e}.'.format(odds, pvalue))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imprinted genes with significant ASE are more likely to show bias in the direction they are imprinted: odds=8.085, p=2.53e-05.\n",
      "\n",
      "74.7% of genes are biased in the correction direction.\n"
     ]
    }
   ],
   "source": [
    "tdf = df[df.p_value < 0.005]\n",
    "pat = tdf.ix[tdf.paternal_percent > tdf.maternal_percent, 'gene_id']\n",
    "mat = tdf.ix[tdf.maternal_percent > tdf.paternal_percent, 'gene_id']\n",
    "pat_pat = sum(igenes_f.ix[pat, 'expressed_allele'] == 'Paternal')\n",
    "mat_mat = sum(igenes_f.ix[mat, 'expressed_allele'] == 'Maternal')\n",
    "pat_mat = sum(igenes_f.ix[pat, 'expressed_allele'] == 'Maternal')\n",
    "mat_pat = sum(igenes_f.ix[mat, 'expressed_allele'] == 'Paternal')\n",
    "\n",
    "odds, pvalue = stats.fisher_exact([[pat_pat, pat_mat], [mat_pat, mat_mat]])\n",
    "print('Imprinted genes with significant ASE are more likely to show bias in '\n",
    "      'the direction they are imprinted: odds={:.3f}, p={:.2e}.'.format(odds, pvalue))\n",
    "p = float(mat_mat + pat_pat) / (mat_mat + pat_pat + mat_pat + pat_mat)\n",
    "print('\\n{:.1f}% of genes are biased in the correction direction.'.format(p * 100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tdf = df[(df.expressed_allele == 'Maternal') | \n",
    "         (df.expressed_allele == 'Paternal')]\n",
    "tdf = tdf[tdf.maternal_percent != tdf.paternal_percent]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "t = locus_p.ix[trio_child.ix[set(tdf.sample_id), 'rna_id'], set(tdf.gene_id), 'major_allele_freq']\n",
    "for i in tdf[tdf.maternal_percent > tdf.paternal_percent].index:\n",
    "    r = trio_child.ix[tdf.ix[i, 'sample_id'], 'rna_id']\n",
    "    t.ix[tdf.ix[i, 'gene_id'], r] = 1 - t.ix[tdf.ix[i, 'gene_id'], r]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tig = igenes.ix[t.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f738ec77690>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAVoAAADtCAYAAAD+6b0PAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAHJdJREFUeJzt3XmU1NWVB/BvVe87NIKAggiSCwYXIgoSIcYFjHGMUXMM\nLjFGEzUxapzRiRtJjEsSEzNEY9QhJnpy3OLoJJKRGDcURNyiIuADQYLsDQ3dTe9dXfNHVZuWQNe9\n3b9X9avq7+ecOtLU7dfXrurbj/e7v/ci8XgcRETkTzTTCRAR5ToWWiIiz1hoiYg8Y6ElIvKMhZaI\nyDMWWiIiz/J9DVxT08C+MepXKsrs85aGxk4PmWSfwYMrIn0d45LIKHXNuSe+ts9fz8JboSUiSqe8\ntJZOGxZaIsoJeZHwVloWWiLKCZzREhF5VhgNb6VloSUileqWtab42uJRXvLYGy4dEPUD7CDILC4d\nEBF5xhktEZFnYb77ioWWiHICZ7RERJ5xjZaIyLN+2d5137ub1LFPv7FBHbv6jWXq2PO+9nl17IpN\n9epYANjV1K6OHbd/lTq2rqlNHdvUFlPHtnXor4gvfuKv6thFD/67OhYAhjz2I3Vs5dnfU8cu3Fmk\njp37ylp17MbtTerYCSMHqmMB4I7jh5ri69pLVHG/eXujesz7H31bHbv/2CHq2H0HlwJYro7/n29M\nVsfuDZcOiIg849IBEZFnnNESEXnGGS0RkWcstEREnnHpgIjIszC3d0XicT8nzvy2epxp4FPd66q4\n0mfnmPJoOuEKU3ymVX3wjDq2/vVX1LEDTz3PlEdt2RhTvNbmy88yxQ/91aNe8qBwCeIom0eHHKyu\nOWdtXc6jbIiIrLh0QETkGS+GERF5xhktEZFnURba3GPZy+G8Z29Tx1adekpv0slZlSW6PRoKtq4y\njbu9QnqTTo+qt+v3DehSO+hwVZz2+9Clvjmzu7NW17xp+4TBx/b5a0ZCvHbgrdBquwis4l+ybWKC\nxg4vefiy+c/zUHLVr3TBB81Qj9tQkmdLpFm/YY3FwbferI7tGOyn84FyU16h8T2eRpzRElFO6Jcz\nWiKidIqy0BIR+RWJhvfUMBZaIsoJnNFm0MCOzaq4jwpsu937VJXfqIqLfPCaeszG8SeYcqhuWauO\nrS0eZRqb/umyv6xQxT14ZvBdEoC9U0LbJZEJXKMNUEenfguF8gb9ETlW508cro59//L31LHTFz7V\nm3RSevi9rerYb+3X4CUHAHg7f7Q+eEccI4t1V5LLhk2wJbJLfxSRVry12RR/zt8HqGO/9rjDXV8c\nb00pY96usBXkIwL4muw6ICLyLBLi3btYaIkoJ0TzeDGMiMgrrtESEXnWLwttVZ7tgkpdrMJTJpTN\npt38vCrurZ/M9JxJ5lUV6C+21bWXeMmhKcTroGFeOvB2wkJb7UbTwLlcaMsrCkzxuxqCvyKejbRF\ntsvLNxznKROdtc32fTVGlejmOpYiC+gLbWWpfszNLfoi+7Bh0yUA+P7nx/a5gr920nHqmnPU/Od5\nwgIRkVVeAdu7iIi84p1hRESe9cuLYURE6RQJ8cUwFloiyglBLR2ISATA3QAOA9AC4CLn3Jpuz58D\n4CoAHQB+55y7J9WY3gqtpYugoiwKS89BQ6PtWI9My7YuAsuVaACob/KTR6a7CKwOHWT/cdJ+75oL\njF057boOCMtrVwp9I9GsQ4fpBw5IgLfgngagyDk3VUQmA7gj+XddbgcwHkATgOUi8rBzrq6nAcM7\n1yYiMojmRdWPFI4BMB8AnHNLAEza7fl3AAwE0NVDl/I3EJcOiCgnRIPbvasSQPcZaoeIRJ1zXf+U\nXgbgTQC7ADzhnKtPmVtQmRERZVIkGlU/UqgHPrGa+XGRFZFDAHwRwAEARgHYV0TOSDUgCy0R5YQA\nlw4WATgZAERkCoCl3Z6rQ2JtttU5FwewFYllhB5x6YCIckKA7V1PAjhRRBYlP75ARGYBKHPOzRWR\n+wAsFJFWAKsB/D7VgP4K7R9m62MvvtlbGtQ/fOGuV0zxT1821VMmlClBFdrkTPXS3f56Zbfn7wVw\nr2XMUMxoG+69ATj3pkynQUm+2rVync/vW1uTfcOaTCo1HDkVFJ6CS0TkWSSPm8oQEXmVVxDechbe\nzIiIDLjXARGRZyy0OejmhWtN8TccM8pLHpk26suG7pKktU9m/sLn1g79xZoh+bp76Oe8sd6Uw+wB\ny03xdQfNMMX3N/3zYhi7CCiNLO1aliJL2YMzWiIiz1hoiYg8C/MpuCy0RJQTomzvIiLyi0sHpDao\nbpkpfnvVp1Vxbfm2N2FhR3adYkH+VTetSR3UTW3paE+Z7Fn/7DrIcbnarmUVhlYtK227ltUVk/Y3\nxdfBFk89i/IWXCIiv7h0QETkGQstEZFn7DogIvKsX85on9uY8mDIj51+QIE6Ntq4XR3bWZryKJ9P\nqOsoU8dWFsfUsdHmnaY8LKobV6viNleNNY17+6vrVHG3HdJuGrd9+WJ1bGTaLNPY9S3BXwyxdIHE\n84vN49eWjVHFRYtt/2+dLbr3Z3mF/mcPxo3NOwrSW/jYdUBqjzWOUMd+papGHTukXR97/ZJGdSz1\nH5GY/pfqlurxHjPZs0iUXQdERH6x0BIReZbtSwciMr2n551zLwWTDhFR7+TCmWE3AvgsgJcAdACY\nBuAjADUA4gCO85IdEZFWfmGmM9grbaFtBXC4c24lAIjICABznXMzvWVGRGSQC10Ho7uKbNJ6AMN6\n+oTX1u5QJ/GVKtuZ9dqWmMp8fQvWtc+vB7BNHz/1AHXsqqYKdezIKmBsme43cy0q1eMu3LJLHbti\n0yZ17GmbgPu/PEEdjyN1rx0AoEUf2pYfNV1x0G6ao920pzcGtm/EwPaNqtjG0pHqcYsXP6KOzT9o\nojo2Hs1X/+xVR1vV4yboWyv3Kgcuhr0hIn8A8AiACIBzATzrLSsiIqscKLTfBHA5gEsANAN4BsBv\nfSVFRGSV9UsHzrlWEfkjgOUA/gpghHOOG5YSUXiEeEar+hUgImcBeArAHADVABaLyLk+EyMiMonm\n6R/pTk0Z958ApgJocM5tBTARwLXesiIiMooUFKgf6aZdo4055xpEBADgnNskIhlbOhgQ26qK68Qg\nz5lk1uwF+qNFZowb4jETvbenH6uO/dTCBerYMNziaNloCABg248nFJrzdKdT6PthAhTipQPt+3OZ\niFwGoEBEDgfwbQBv9/QJt8w8SJ1EG4BdDbp3nbbIAomdvnbm6QrMxX+ZrR4XADD1AXXoUYX6tjEA\n2IHhtlwUnnl/K276nO4Mp2Ms7Voh0dEZR2lnPNNpmPxylb4wjBhQqx948AwcPzz4UqctsgCwusn2\nq+/gANIN86Yy2qWD7wDYD4mOg/sB1CNRbImIwiEa1T/STPtr5y7n3AXguiwRhVQuzGgniEi510yI\niPoixF0H2hltHMA6EXFILB8AAJxz3EyGiMIhW29YEJHpyS0Qr/adSFm5suWizjbuwI7NqjjDpQYA\nQEmZYbHfeJLNoB3vKiP9/COjotx2IaNhl22vikyrjesvmh2Yr98jIo7S3qTjhfZ9DwA78od6zCR9\nIgXh3b0rEu/hTSciS51zh4jIa865oywD19Q0qN/N6iILIGb4ISmpW68ft9L2ZmuP6ItRPvSdcIXr\ne2zm+IT64foNQQCgvVnXfsRC+0+WQvvUBlseJ4y2nWnX1qT7PluKLOCn0FYW2XrXiiqr9S0NexF7\n/2X1C5s3blqfv55Fqp+ojSKyHsA+ItK9aTMCIO6c0/ULERF5ls17HXwBwP5I3H57qv90iIh6KcRd\nBz0W2uTGMesAHLa3GBF5yzn3maATIyIyiWTvjFYjrWsdRER7lOOFNrvue6TA3TjkEFP8j7cuDTyH\nW4cdqo69ZOM7gX/9/uDn++31H7b/4qZtb3jMZM/iOV5o+6xxl5/dNWaPmKGObTPeJ//DLcEXCwDA\nQH3BuLHiYNPQ2gJ3Tbl+XOv3LQzuGa4vGABw3SZdu91048X7Gyttr9+Nm3V5XH+A7Si//9gQ/C+e\n2ftMMo07OIgvmq1rtEREWSOgrgMRiQC4G4lrUy0ALnLOren2/JEAfpH8cDOAc51zbT2mFkBeXKMl\nooyLR6LqRwqnAShyzk1FYn+XO3Z7/j4AX3fOTQcwH0DKk1rVM1oRKQMwBsBSAKXOucbkU7dpxyAi\n8ia4NdpjkCigcM4tEZFJXU+IyKcAbAdwlYhMADDPObcq1YDao2yOB/AOgD8BGApgrYjMSCbymPX/\ngogocJGo/tGzSnzyZv8OEen6pH0AHA3gVwBOAHCCiBybakDtjPZWJKr808nTFT4H4GEkTsPdo44C\n/W+XikJ97Ac7rOfF+/F+Q49LMv9iXIXuPuy8Yn8L+pUhuRV/U7vuluT8aDhWpQpL9ZcytLfJkgfB\nzWjrAVR0+zja7TDa7QA+cM6tBAARmQ9gEoAXexpQ+w6KOuc2dzvKZnnXn8Psutpltk9o1R9FYi20\nPtxavxyxFl3OliJ7Xe176tiqItsvhlW1fn5R/rBuuTq2KKLfe6I17q9l6Ce7Vpjitd05ProIAOD8\ndfp9ODIhHg3s2v4iAKcAeFxEpiCxXNplDYByERmdvEA2DcDcVANqM1svIqcAiIvIACROXFhnSp2I\nyKdIYP8CehLAiSKyKPnxBSIyC0CZc26uiFwI4OHkZPMV59zTqQbUFtqLkThqfAQSFf05AN+yZk9E\n5E1ASwfOuTiAS3f765Xdnn8RwGTLmKpCmzxifJZlYCKidMraO8NE5EP0cIstt0kkotDI4m0Sj01H\nEkREfRbiGW2PJyx0JyJnA/g0gFsAnOmce7Cn+PYtH6pvgt9WNEQbCgDIV7YHrW22tdqMKtFftXxn\nZ4s6dtKwMlMe2pMQfFm4RX+qwMkji0xjb4vp4/fJt+2BUd8S3nvdw2pxTWPqoG6OHqx7L1u7cqaN\nHtTnK1ltO7eqa07hgCFp7R3U3rDwEwAnAzgdiVnwBSLyi54/i4gofeLRfPUj3bRz7ZkAzgPQ4pyr\nB3AiEqcvEBGFQ3B3hgVOW9q7/q3eNTUv6vZ3RESZF1wfbeC0pf0xAI8CqBaRKwG8BOAhb1kREVll\n+4zWOfdTEZkJ4B8ARgL4gXNuntfMiIgMsrmPdnq3D5uROA334+eccy/5SoyIwqeqoFkZmYEOkBAX\n2h7bu0TkhR4+N+6cO25vT9bUNGTfGScEAKgstrWXsa2qf9AX2YS69hJ17ODBFX1eYG1uaVHXnJLi\n4rQu6KY6bvzz6UqEiKgvYiE+vy7V0sEL6PkW3L3OaImI0im8ZTb1xbAfpiMJIqK+CvGENuXSwYKu\nP4vIRADlSBzGmAfgQAAL9vKpRERppd1OIBNU7V0i8gCAqQCqAawAcDgSu5Dfv7fPGRDbakpkZ55t\nv4OgPTRmUuqgbs5e/YanTPQsxwW9Pl6/febMlQt7k07g/nawactPnLh8iadMdFrvusr8OUWX7X7A\nat/NH3+UKf6kFa8FnkMmZO2MtpvpAD4F4E4kDiWLALjLV1KUWe9faNt6ePiveT5nf/Du2bb3xQEP\n/K+nTPYsxHVWfWfYRudcOxKz2UOdc8vwycPLiIgyqjOuf6Sbdka7QUSuBfAsgJ8lz8op95YVEZFR\nLMRrtNoZ7YUAPnTOvQ7gCSSOtdn9TB0iooyJx/WPdNPuddAA4JHkn+9EYq0WACAibznnPuMnPSIi\nnVy4GNaTQG5lq1qZ8sReAMB90SP1YxYV9DadlAa2rVfHPrhJf8LCOeOq1LG1HeG4t3v+R3Xq2FnD\nbLdxWgxsXaeO3VE0Ujdm+0b1mJvVkf9UuezPqri8qkG9GF1nQOc2Vdw/jONW71qlDx7c97la1rd3\npbDH/7vIltXqATp37Qwgjb6pfvE5nDRCX+RgKLS+vGIocF+a91PT2DtGTk8dBFuRtTrrdf1OnPF8\nf79UtYaeepopPtbg530/62+/xI79P6uK1RZZAJj48xv1SZRU6mMDEuYNstN/pgMRkQchntCy0BJR\nbugMcaUNzRotEVFfxMJbZ9Wn4P6uh6dvCygXIqJey/r2LgATRKTcObdr9yecczlz/2Ws0HAV33Zs\nfdYpfXaOLlC+7jUPXza16y6dDPSchy9Vha26wBa/eaRTZ4hvwu3xhIUuIrIEwFgADokjbQD0vB9t\na6P+hIW8nfor+L9eW6SOnX77JerYYQ88qY4FgEEtxk1zovuo4t5vsFXwcRWFpngNdZEFEMm3ff1l\nk7+pjh1brX+tNzR0qGMtdxCNj+qvyq+MD1bHAsD59yxWx8676hh1bDX0LXSxBfrOjtgJ+tcOAJob\n9a9JECcsvL+lXv3Cjtu3MjwnLHRzjdcsiIj6KMw3LKj+rZzcl7YUwL8B+DKAAd33qiUiyrQwr9Fq\nL4Zdg8RpC+sAfAjgehG5zmNeREQmsXhc/Ug37dLBuQAmO+eaAUBE/hvAmwBu9ZUYEZFFLvTRRruK\nbFILAP1KNxGRZ7EQ34Or7TqYA2B/AL9P/tX5ADY4567Y2+fU1Oi7Dih7VS61dWvE22z9RA1H2Hb1\nz1WVpbb4+iY/efgSRNfB6+t2qGvOkSMHhrLr4EoAlwD4GhLrus8DuNdXUkREVmHe+LvHQisiQ51z\nmwGMAPCX5KPLcBFpdM5t95kgEZFGNq/RzgVwChLHisfxr/saVIjIC865r/hIjohIK8xrtD0WWufc\nKcn/Hrin50UkAmBt8GkREdm0d4a30qrWaCVxGuO3kTiQMQIgD8CBzrnpIjLWY35ERCphvjNMezHs\nUQB/AjANic6DLwB4DwCcc3u8Ob98yYPqJPKmfEkdW5On3+bj6VX6+9S/Ub1JHQsA319hO2396im6\no1O2dujfLQvX7TDlcProalXcgLUvqMf0PYeoymtQx9bFdK+J5Qq+ZR+OJ7YbTugAcEZVjTq2o3S0\naewl2/RtB5P30X1DLK/FjrnGTf2uuTN1TAqxEFdaSx/tD0SkAMBbSHQcvOIvLcoW2yaeYYpfs0Pf\n3jV5gHIHqn4g0qovcs9syfOYSXiF+WKYdl/AJhEpArASwBHOuVYAxf7SIiKyicX1j3TTzmj/AOAp\nAOcAWCwiJwHI/OmERERJQc1okxf57wZwGBJ3wV7knFuzh7h7AWx3zqXc90U7o30UwDwAlwJ4EcBw\nAO8oP5eIyLtYZ1z9SOE0AEXOuakArgVwx+4BInIxgAna3LSF9v+QqO4RAB8BeAJAo/aLEBH51t4Z\nVz9SOAbAfABwzi0BMKn7kyJyNIAjYbg7Vn04o3PuQm0sEVG6BXgxrBJAXbePO0Qk6pzrFJGhAH6A\nxKz3LO2A2k1lrgewBYk9Dj7etcs5t25vn8NNZfqH1jz73hxFYT6ulDIiiE1lHvr7evUb6+yJ++/1\n64nILwAsds49nvx4nXNuZPLP30Viz5cGAMMAlACY7ZzrsZ9VO6OtAvB9AN0bU+MAbM19RESeBPj7\nexESWw88LiJTACztesI5dyeAOwFARM4HIKmKLKAvtGcAGLLbnrRERKER4NLBkwBOFJFFyY8vEJFZ\nAMqcc3N7M6C20K5B4uRlFloiCqWgtkl0zsWR6LDqbuUe4h7QjqkttHEAy0XkPQAf33Lb03HjRETp\n1JkDt+De4jULgxrjQszgXlysyaRVjXvcOmKvxpYVesqEfDvq+r+a4l+7ZaanTHKDom0rY1SFlkeL\n096wg4DCImtPWCAiyha5sHsXEVGosdASEXnGQktE5BkLbQ66ccghpvgfb12aOqgXFh96tDr26HcX\ne8nBly1XftUUv+9/PeIpE7J6aMyk1EHdXFHv+vw1WWgDlG3tWlYPHniEOvbkoeUeMyHf2K4VrLaO\nLD+ckYgo7DijJSLyjIWWiMgzFloiIs86WGiJiPwK84xWdcJCb8Q+WmoauLZ4lCquunG1KY/asjGm\neK2BHZvVse1L5qljC8fp22IihSXq2M66bamDusYtKFDHAkDtvkepY6sidamDuvLo7Egd1MXwPu4s\nqdKPC6C+Vff9qPrgGdO40cJifeyAwaaxt1eOV8VV5TWox8zfoT/4OlZuy7dg3wP73E506ePvqN8E\nvznzsLS2L3FGS0Q5IcwzWhZaIsoJLLRERJ6x0BIReRbr5J1hRERecUabBZ7bWG+KP3OIp0SIqFf6\nZaHVtmuZx/XUrmW1I3+oPvizF/lLRKtCMp0BAKAubmitsjTgWGJbDbEGdQfN8DOwR3WxCn2wsmWs\nN2zNYHvWyk1liIj86pczWiKidGKhJSLyjIWWiMizfllov/rw26b4R2YdroqzdAeMrNLfS27VMudK\ndeyLp/9IHfvh9kZTHpcePlwVZ7kP/5b6T6tjN9W1qGMBYMHLa03xC2cfr4rbfPlZ6jHlzvvUsZHV\nb6hjAWDnaF2+VhXl+h/Vbc0xdez8D7arY78xQL/XwauF49SxADA4gKth8f5YaLNNUV400ykQUR90\nstASEfnVGWN7FxGRV5zREhF5Fg/vhJaFlohyg69DDILAQktEOSHMSwfejrKpqWkI7/81EXm134wr\nTPFtf7+/z0fLTLn5WXXNefWGE3iUDRGRFftoiYg8i7G9i4jIL85oiYg8C/PFsKwrtG35tltlCz1t\nBry6qd0UP6a0wEselLDd+EM2KKq7FtKaZ7tmUhTz88Melvd9mLG9i4j6lQ3PzEn71+QNC0REnnHp\ngIjIM14MIyLyjO1dRESecUZLRORZUIVWRCIA7gZwGIAWABc559Z0e34WgCsAtANY6pz7dqoxs67Q\nhqVtxdKuVV5ha+3a1WBrHSN9u5aVr3YtAHhoZY069sKJw0xjN2f458Ta/hjASTZBXgw7DUCRc26q\niEwGcEfy7yAixQBuAjDBOdcqIg+JyCnOuXk9DcjzW4goJ8TjcfUjhWMAzAcA59wSAJO6PdcKYKpz\nrjX5cT4Ss94eZd2MlohoTwJco60EUNft4w4RiTrnOp1zcQA1ACAi3wVQ5px7NtWALLRElBMCXDqo\nB1DR7eOoc+7jtZjkGu7PAIwFcLpmQBZaIsoJnR1tQQ21CMApAB4XkSkAlu72/H0Amp1zp2kHZKEl\nopwQ74wFNdSTAE4UkUXJjy9IdhqUAXgTwAUAXhaRFwDEAcxxzv2ppwGzrtDOXrAmdVA3N31utCou\n1zsDisv0L3W78Up7rCWwN3halJXbXuvGXdn1WgNAieH1bm7sUMVVFhte5yZ9aFDisWDeh8l12Et3\n++uV3f5srptZV2izUbYVZUqP7xzQrI7tMHSvdYSgb39idRT1LXlp/ZoBzmgDx0JLRDmBhZaIyDMW\nWiIizwLsOggcCy0R5YTO/jijbTYeAZJtKksNwYY3wKIttgtnh1QVm+J9sFyNrmnXv+Uue+I9dezl\nx45Rx04xdh1UFulfk/pWHlnUZXFNozr21MEVqYNS4NJBgG6eOdYU36lsPVr0UYNp3GnDM1/gLIra\n9P9/RQDq2ktUsaaWn5B4dUOD+heUpcharbj6anVs9d1/VMd+8fYFpjz+9r1pqjhLF4GlyAaFhZaI\nyLOg+mh9YKElopzAGS0RkWcstEREnnV2hPcOzJwvtBvawvtbjjKnLnUIgMTGpBYdBdm3l752n4/8\nnRsMo5b1Lpk+CPOMNqLYbbxX6hqb1QNbjt0oN7yRP9jRmjqom/0K/dybbdmwprnddgSJdkOXncaX\neYCyOy9abPueleXr2/6aDDftl0X0s5mPmvXvoaGGzVka2myvXb71+J3W4AuJ5b1pKbTbioea8hg2\noKzP/aAV0/9D/YZpeOnnae0/zfkZLRH1D/3yhgUionRiexcRkWdhXqNloSWinMBCS0TkWZh37/LW\ndUBERAnZ1/RHRJRlWGiJiDxjoSUi8oyFlojIMxZaIiLPWGiJiDz7f0euCIZ6buSTAAAAAElFTkSu\nQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f738ec77150>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.heatmap(t.ix[tig[tig.expressed_allele == 'Paternal'].index], xticklabels=[], yticklabels=[], cmap='RdBu_r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f738ebe8910>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAVoAAADtCAYAAAD+6b0PAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xl43WWZN/DvOdmafWug+0bLXbaWylZKKQq2bAVbRBRB\noVIVwXmdcUAHGctLXwW1yMAovLKoOAOCiBYQpINgKVAWKUsplD4tBChdQmP2fTnnzB8n4Yqlzbnv\n5DzJ75x8P9eVS8v55unTLHd+edZQLBYDERH5Ex7uDhARpTsWWiIiz1hoiYg8Y6ElIvKMhZaIyDMW\nWiIizzJ9NRyt3KBeNxbNzjW1XZczSZUridWo29wSKTX1ob6tW52dF3pfnY0WjLb1A7Z+D7c/Vtaq\ns5eMbza1rf266MzUP180d0VMfSgLhUx5H0o7tquzscxR6my4q1Xfbu1udRYAMmcvGvQH7tLQFHXN\n+UXsvSH9RHkrtEREQylj+H/G7RcLLRGlhYwA/DaxPyy0RJQW+ERLNELM+ef7TflXbzrPU09Gnuxw\ncCutt0JbUyi+mlarD5Wrs2OMH4kxhdnqbC1m2BpPY+dMK1Nn66DPWhz++Z+Y8lvvuUKVsxZZX7ST\ngmYZhuy4KaamK0zpfePQARGRZxw6ICLyjE+0RESeBXn3lbdCO3/lk6b8sytO8dQTSlXHXPVnU/6l\n68/w1BNKBXyiJQoQ7eSWFVcQDC+O0RIReTYil3cREQ0lDh0QEXnGoQOFqu6oKnfudWvVbb4y7w11\n9snZy9VZAGg1nOr05Yu/b2p75+M3q3Ivz12gbvPByjp19ua1P1RnZ/0hX50FgI2nW051sm1YKA41\nqHLd+foT0iYs1W9uWHnNl9RZAJhUrD+17uQpxers6be8oM5uf0v/+bjmsvnq7BKxnUKXDCPyifbN\nqw5RZ98KjfHVDUphr910ljqb164/fpHSE59oiYg8Y6ElIvJsRA4dEBENJS7vIpOijQ8MdxeIUk6Q\nhw5CsZj6mh2TzvX3mxpuOPh0L/3Q6jB+lnIifj5u1iLbOPtcVW7L2aeZ2p358BpT3oeiTatN+cYj\nlnrqSfra3Nihzp6Qq7+Dr/mR35j6UfK16wZdJh848DD1N+W5H77JO8OIiKyC/ETLQktEaYGTYURE\nnoVZaJNnxbpKU37lSdM89UTPOtYYBMVbHlVnG2ae6bEnRDqhAI8deCu0wz25ZeVrcsssnOFlUuf4\nX1yjzkYbg7HLipNb/h1alKPOfpg5QZ0dPX3mQLozKBnZlkvNhlbKPdESEe3LiHyiJSIaSmEWWiIi\nv0Lh4N4aFohCm1+Q5a3t0q5d6mxd1jh1tihXd6zjSFBSqb8fLnzgFHV2bdd4Uz9ml4xS5Uqif9c3\napjJfqu7VN9uj7FZuuJQnNliarehW3d8peV7r7NdfzTocOATbRJddsIUdfaQWJW/jhhkTThouLuA\nXRPnmfK5yslBS5Gl1Kb8mQAA6Nr9nqlt/ZTc/nGMlojIM646ICLyLJSk07tEJATgVgCzAbQDWO6c\nq+zz+gUAvg2gG8CvnXO/SNRmcEePiYgMwhlh9VsCSwDkOOfmAbgKwI17vb4KwMkA5gP4VxFJeM8Q\nCy0RpYVQRkj9lsB8AGsAwDn3IoCj93p9I4BSAL2XviWc0AjE0EFLc5c6OyZT/7OhDvpVBFaNbYaf\nUaWzvPVDSzu5ZVU/7RQv7QLx39t8qA/7uThwrL/FMwi3N9neIVO36sDyvWf5xbxt0b8Y0kCBKb1v\nSZwMKwLQ96bPbhEJO+d6lxq9CeBlAM0A/uica0zUIJ9oiSgtJHHooBFAYd+me4usiBwB4EwAkwFM\nAXCgiHw2Yd8G9C8iIgqYJA4drAdwBgCIyFwAm/q81gCgFUCHcy4GYA/iwwj9CsTQARHRYGVkJW15\n12oAC0Vkfc+fl4nI+QDynXN3isjtAJ4VkQ4A7wC4K1GD3q6yqa5uCshxWDpLf7XBlL/hC7YRxIPy\nPA7gEQrz9b+cZdbvMLVt2TFosbNTv9Pq8HC1qe2nmkvUWe2OOp8qKgoHPcD66pJF6poz58HHeZUN\nEf2jN6IVGK9ckL+xvt1zb4KJO8OIiDwLJZ7kGjYstESUFnioDBGRZ8naguuDt0K76KZn1dnHvrPA\n1HZGp+6IwuqgXE8DoHjrY+qsj2uArvzLNlN+1cIZqlxZq+0ONwCozRv+e9wsNjd2qHI5hs00ADDK\nmB9u5U1OnY0UHmBsvTBxJAHF+thhwyfaHrdfdBQqPP3qYSmyNDBNLfrzgXdjjDpb16bfOWU1s9S2\nEkX7bwzCKoLhEObpXUREfvGGBSIizzh0QETkGZd3Ucr51mNbVLn/Pinbc090NhwzX50d/9zTHnuS\nWi7902Z19vefDO4YKDBCC+0TY/+qzn74T7ea2q746W9Vueuf2Gpq98ZTxZTXOvnZClN+9cHJ78MZ\nh+sngB58bac6+1xkgrkvM83vkVx/21Fvyn9mapkqV29c5PJBs+2Cz5JhXr3U/upT6uzZu+ea2n7i\nm5OMvfk4jtESEXkWygjuEzcLLRGlhYys4Jaz4PaMiMhgRI7REqWLNYccq8rN3fw3zz2h/gS50PI8\nWqJ+aItsr9PeYrEdiGScR7vz2q+ra874a27jebRERFZBfqJloSWitMBCS0TkGbfgJtGGmlZT/ujy\nPC/9eKW2zZT/RFmul35oNV33dVO+8Hu3qXKdxqP+ytbdrs5mzznZ1Lb2+MXw71aa2rUoa3lHnW1+\n8gFT251nf9fanYQsd60BtlPShlqYy7uIUtP5T96szoaKrWewUjJx6ICIyDNuwSUi8izMLbhERH6N\nyKGDok2rTfnGI5Z66ceWpk5V7tAKz5NVd69QxZq/pJ+omVz7ujrbpE5SUEWydYXkgOpN6jY7848w\n9SHnkVWqXN6s40ztouJMW34fRmSh9eWUmnXq7ItjT/HWj+PGF6izXXdc7a0fWpMuvtiUr1PmLKsI\nejWfsFyVW/XCdlO7VypP5mv54jXqNndc+jlTH8bdcr8+bFhFoC2yVtm7NqGm+DBVVltkhwtXHRAR\necYnWiIiz7jqgIjIs1CYqw4oxRQUZnlrO/+vP9cF8842tVve5FS5HSXDfZkOeRHgQuvtmMSqhhZT\nwxmdwd3aty/1K5aZ8iUrf+2pJzp7uvWfjmml+gsXw522LdGRx3Rbe3u1nPxNVU5bZHvVFPq5H86X\nv+xoMOUXTihOeh/KG9405bWTbEByjkls/cMN6i/yvM9eEbxjEkVkQX+vO+d4rSgRDat0uDPs+wBO\nAPA0gG4AJwL4AEA1gBgA2+kfRETJlqn/TWyoaQttB4AjnXNbAUBEJgK40zl3qreeEREZpMOqg2m9\nRbbHDgBj+3sHy5jr2XfYrv94+Ku260V8GO4xV6sDMvVDUs1NXYaWjZNmyjFXqyCMuVZH7PMdFRm6\nz4uPMVcry5jrsAjwZJi20G4QkbsB3AcgBOBCAE946xURkVUaFNqvAvg/AC4F0AbgcQC/9NUpIiKr\nlB86cM51iMjvAWwG8D8AJjrnUms9FhGlt1R/ohWRzwP4dwC5AOYBeF5ErnDO3e2zc4NVNCqizlZ1\n2vZu5EVT6zb1jfXtpvzsklGeejL8gnKyXBAU5VjG44HGDn8bWQYt1QstgO8iXmCfds7tEZE5iI/R\nJqXQBmFyi2iwtBNb5EcoK7g/BLSFNuKcaxKJz+w653aLCIcOiCg4kvREKyIhALcCmA2gHcBy51xl\nn9ePAfDTnj9WAbjQOdfvwdfa0eM3ReSbALJE5EgRuR3Aa9Z/ABGRL6FwhvotgSUAcpxz8wBcBeDG\nvV6/HcDFzrkFANYAmJyoQW2hvRzAeMRXHPwKQCOAy5TvS0TkXzisf+vffMQLKJxzLwI4uvcFETkY\nQA2Ab4vIUwDKnHPbEjWoHTr4uXNuGeLVXaWqWz+yMOPV36qzmWOnqrOR6coj9wfgqw/ZDti44zO6\nxd6lHfpbBULNtfoOZNhOrMp++MeqXP5pF5ja/c4G/QTlxcdMVGcPrX5RndX3IG7RTc+qcmsnP2Nq\nd8upV6iztzxTmTjU41eLRquzURSps4Dh5DXoDwRKliQek1gEoO8pPt0iEu5ZaTUawPGIP2hWAnhE\nRDY4557qr0FtoT1cRAqcc80D6PSw+aBd/4HPCKXWKgKrE966V53trN7jsSfD7/XpZ6mzX7vtBY89\nGX7h9kbUh8pVWUuRHRbJK7SNAAr7ttxnOWsNgLf7HEewBvEn3qf6a1BbaGMAtouIQ3z4AADgnONh\nMkQUDMnbsLAewGIAD4jIXAB9b7usBFAgItN6JshOBHBnogb7LbQisqDnCMQrB95nIiL/QllJO71r\nNYCFIrK+58/LROR8APnOuTtF5BIA9/aswnrOOfdYogYTPdHeAuAIAKucc1zsSkTBlaShA+dcDMA3\n9vrPW/u8/hQA033qiQrtLhHZAWC0iPQdjQ8BiDnnpln+spGs+O3HdcGJvGaFhlbu4/+hC2YGd0MA\nEOyzDvq9ykZEwgAmAPgTgI9d4OSce39/71td3aSeXSqJ2CZf6jMO0AVz9D/hyrY/b+pD7YH6B3x1\nke3RMH2RKlfW/p66zVjtblMf6sYdb8r7UF73ujobLVR+TfSoyxxj7U5CsWz7N3rIwxVOr9S2JQ71\nOGHDL0xtty36F2t3VJJxlU307RfUNSc8fW5wrrLpmWnbjvgOiX0SkVecc59IdseIiExCwX2iTcYt\nuNzgTUTDL80LbXovQCWilBBL80JLI5j1+MWT6g1j4cW6hfQjwYp1+p1hS47o95apIVHW8o7tHSqO\nHPxfmgbHJHqlntyy6tBvtrRMbllpJ7esakdN0YfHGbIBEWmoQf2UT6myV6+1fWP/UNesiY+JrYF4\ncNNurDxJtyDorOpPmtq+fwD9GTIBXnXAMVoiSgtpMXQgIvkADkJ8O1qec66l56XrfXSMiMgkwIVW\n1TMROQXARgAPARgD4D0RWQQAzrlA/zZBRCNEKKx/G2LaJ9rrED+j8bGe2xVOAnAv4rfh7lPx1oTb\nfz+SMf5gdfZLT+vvOLprqn7cLnPsFHUWAJqf+bM6+/7jG9TZ2E/1p2xlr7hInZ1+rf4Xj43R4Z9M\n6bV2d5OXdos2PqDKbbld/xwx+dNzTH0oOMN2xKSFr4/b9q+co8qV/WyVl7+/XwF+otUW2rBzrqrP\nVTabe/8/jWwnHmic6S3Rzy6v3q2fQrjlkyXqbPT1tepsUHzvtf9UZ58/7/+ps2sO1z+MvHGz/tzo\ndWd/TZ0FgJNf/5spvy+xcCDm9vdJ27MdIrIYQExEShC/cUF/QjURkW+h4M7La5+1vw7gAgATET+P\n8UgAth9ZREQ+pfoYrXNuD4DzPfeFiGjAUnZ5l4i8i3622PKYRCIKjABvWEh0TGK/1+j2d0zijtpm\n9RkI7Su+oo1i8oqfqLPPNOarsx2GyyQB4JQ63YV9vbS7wyxHKv46fHTiUI9DKwrU2WMrH1Zns6Yd\noc4CQOeWl9TZHcdcqM7+8m8fmPrx7WN1Fz92Zuq/ectfus/Uh+yD9B+7aF6pqW3tMZCW1UHVh55p\n6kO24XsqGcckdtZVqWtOdumYQB2T+FEhFZEvAjgMwA8BnOuc+y/PfSMi0gvw0IF2w8KPAJwB4BzE\ni/MyEfmpz44REVnEwpnqt6Gm/RFwKoAvAWh3zjUCWAjgdG+9IiKySvVVBwB6B1t6x0By+vw3UlCP\nhQX4qLd08k6rbofhxKIczz1JX0W5Q1wiAryOtt/JsF4i8l0ARwE4FsBNiD/d/sE5d93+3qe2qdV0\nIHikXXekoXbrZK/G2eea8lp/2dGgzp7b+pw6u3vmGaZ+5EV1H+Ym49dgofKzt6Wp09TuzNG5to4Y\njrrU0hZZAMjPtv3gG2OYPEtn1iKbU1A86CrZ0aK/pzAnf/CTbxbadbQ/FpFTAbwPYBKAa5xzj3jt\nGRGRQSqvo13Q549tiN+G+9FrzrmnfXWMiMgkVQstgGv7eS0G4OQk9oWIaMBiAR6jTbSO1sOFH0R6\nM87/sTq77d7veuwJBV1EOV8xHBLtDFuL/rfg7veJtrpaPzBNtC+WIgvoC21p1y51m8ufakkc6mPV\nwhmmPMUlY2dYU2ubuuYU5uUGajLs/w5FJ4iIBivAD7QJhw7W9f5/EZkDoADxyxgzAEwFsG4/70pE\nNKQ0S1WHi2p5l4j8BsA8AGUA3kL8PNr1AH7lr2tERHop+0TbxwIABwP4GYD/RPyp9ue+OtWfvCdu\nNuVbP/0tVW5zY4ep3XE3XKrOfmXa5ersLRd8wtSPsVm6JS0r1lWa2l15kp8TMDtvvcKQrvDSB592\ndto2WIw3bohINsvpXbEj9bvuw017jD0pNOY/LsB1Vl1odznnukTkLQCznHP3icjgPzJE/XjzxGpk\nX3ZD0tutyxqnzq5aqG/XWmQpudLhiXaniFwF4AkAP+m5mFF/wCkRkWeRAI/RardSXALgXefcSwD+\niPi1Nt/w1isiIqNYTP821LRnHTQBuK/n//8M8bFaAICIvOKcsw0sEhElWToMHfRnnwt/i3L0JyQB\nQGNHVhK6QkQjVZCXd6mOSezP/p5oOxprTQ2z0BKNXMnYGbbdcE/hpLKCQO0MIyJKCQF+oGWhJaL0\nEA1wpfU2RktENJQiwa2z6i24v3bOLdvPy9cnsT8JlcRqTPn6ULmXfliu1Mmaepip7ZqiQ6zdGTbh\n36005TOX2fKdrd2qnOXrItymv4YoUmjbndYQ8bOPp7zhTVO+ptj2NZdsZTWv2d6h4sRB/53JeqAV\nkRCAWwHMBtAOYLlz7mNbK0XkNgA1zrnvJWpTu472cBHZ5wYF59z9+/rvjR1ZpjciosGIIqZ+S2AJ\ngBzn3DwAVwG4ce+AiHwdwOHavlluwd0uIg7xK20A9H8eLRHRUEriEO18AGsAwDn3oogc3fdFETke\nwDEAbgMwU9OgttB+x9BJIqIhl8QNC0UA+o4vdYtI2DkXFZExAK5B/Kn389oGVUMHPefS5gE4C8BS\nACV9z6olIhpuSdyC24h/PE4s7JzrvT/9cwDKAfwZwL8B+KKIfDlRg9rJsO8A+CyAexBfZXC1iBzm\nnLtuf++T//RtmqbjnTj5QnUWtg1nKN7yqCqXMVH1G8BHdFM0AxMepTs6L+PeH5ja7Vp6lSqX+/h/\nqNu0HS5pV1L5pC449Ui/HVEqa9UfRxnN8Xcuk/r4w7DtmMaG6YsG0JuhkcRDZdYDWAzgARGZC2BT\n7wt9jyAQkYsAiHPuvxI1qB06uBDAcc65tp6/4A4ALwPYb6H1JZaZrQ+/6e829PAs/b2V0W59OWoo\nnDiQ7gyb4iUX296hcYc6Gtuun2nPaP67OhttqlVnW8v1d4AV1mxTZ83C+pWY3bttZw/7sKNijik/\nOQl/ZxLX0a4GsFBE1vf8eZmInA8g3zl350Aa1H72wr1Ftkc7/D7UERGZRKKJMxrOuRg+fjrh1n3k\nfqNtU1tonxSRPwC4q+fPFwH4q/YvISLyLR12hv0zgEsBfBnxCbS/Ir60gYgoEIJ88He/hVZExjjn\nqgBMBPBoz1uvcSLS4pyzbdUiIvIgyE+0/R6TKCKPOOcWi8i7iN99tve5BoUA1jrnPrf3+1ZXN6n/\n1foNkcCjW6vV2UtnlamzWbuNWxxLZ6mz2849U5099r5fmvpRlznGlNco7diuzlbl2acxcjxsSrdc\n2tm++NvqbMnOl9XZSJ3tQsLMcv3nLtpUr89OP07fh+q31VlAv7VXu9qnV/aJXxj0mSnPVNaov7BO\nnFYenGMSnXOLe/536r5e79kT/F7yu0VEZNMVTdJsmAfadbQC4DLEL2QMAcgAMNU5t0BE9OtfiIg8\nCfJVNtpDZX4HoB7AHACvATgAwBsA4Jzr9NM1IiK9SDSmfhtqlnW014hIFoBXEF9x8Fx/79BkGQEJ\n8E+i/dndFYxfU3LzdZ/CrIhhD5fn7V4dGfovDh/juTRwpd1VqtxwfHcEeTJMW2hbRSQH8UW7Rznn\nnhWRUcnqRDgEFCo/RmW5+p1h929rxmkTi3V9GGvbyYIm/V7go3/0LX27e95VR9tnDP4Mz32JFI9H\nY7tua+a42jdMbe8qU58sZ9L6af3HWPtrHADUjz8K0faIKltarCtCvWKdrepsKKcAtXnTVNkywwQe\nsvXfxtF8/dnO9bPO0vcB8V+RByvIP5O1hfZuAH8CcAGA50XkNAD6vZRERJ6lwxPt7xB/EPgGgKcQ\nP4vxfzz1iYjIbDjGXrW0hfbPAF4HsB3ABz1vRESB0ZUGhRbOuUt8doSCpbSTI0OUWoI8dNDvzrBe\nInI1gA8RP+Pgo1O7nHP73UJk2RlGwWItsnXZE9TZ4rcfN7Ud5PNPKXkqKgoHvVPrt6/uUNecL86Z\nEJydYX0UI36aeN9DP2MAdNOgRESepcOqg88COGCvM2mJiAIjyEMH2kJbCaAUfW7AJSIKkpQ9JrGP\nGIDNIvIGgI+23KbTdeOlu5435evGHe+lH2XNtutQagt41AQA5DyyypTvWHylp5740Z2l32ZR0aI/\nfQ2wjbEHWTQNVh380GsvKFB8fuPddcbVpvzSrZwMI52UX97Fq8WJKOjSYeiAiCjQ0mFnGBFRoLHQ\nEo0A5XWvm9/HciUS9Y+FNgX4WkVgle6rCJZufclLu6m2isAq03D+cbqsIrBioSUi8oyFlojIs87u\nYNx6si8stESUFkbkE21xln63bkbDbnX2/+8qUmeXztRfkJGfZTvM5yFXo85+/pAydTb88p9M/WiY\neaYqt6db/0U4s/1tdTYUsz1FPNA8Tp09p6JFna295+fqbPnic9XZh7unq7Of0d+y9JGy9vdUuVhI\nvzOsKm+yOjum9X11trlkqjqb192szsYVGvMfNyILLdFIU2+8d664/h1PPRmZWGiJiDzrZqElIvKL\nT7RERJ51RrjqgDwpfPV+VW7PEZ/z3BOi4RXkJ1rVnWEDYbkzzLJCYWd3rjqbf9e/q7Ndl1ynzgJA\nxeZH1dktU081tT1WefaotsgCQCgzS53NHDNJne2efJQ6CwB4cbU6GsrST+PHjjzN1I3G9gxTXqN0\n+9Pm96mbtECV29TQrm5zUvEodXaK4X646zbrVz5cPneiOgsA5YV5g77D65L7XlXXnF9+YU4g7wwj\nIgq0ID/RstASUVqIRDlGS0Tk1Yh8oi2J/j1xqEcM+b66kZKy8/jzz7cS1Kly4epKdZsD+TYv7dql\nTOp3F45UQS603ibDuj5819RwfXi0KldqGLwH/B0Z1xrWj6WPDTWqs205peps4YdvqrPbCg5WZ6fk\n6/9tWbv1fQCA7vIp6my4Rb/NGQDqcnSTeNoiC9gKbbRcv/UVAEKRzsShPuqy9NuXtYpDDaZ8Q6zY\nS7vZoycMenJqyZ0vqGvOg8vncjKMiMgqyE+0LLRElBZYaImIPEtWoRWREIBbAcwG0A5guXOuss/r\n5wP4FoAuAJucc5clajMwhbbkvbW64DjbVS9tGfqhmNxIcH8iUmKlHdtVuVjO4I/kGw6xbP2GgVCn\nn6VO2knuWIZ+g0yyJPGJdgmAHOfcPBE5DsCNPf8NIjIKwEoAhzvnOkTktyKy2Dn3SH8Neiu0oapt\n6mysU7/rJbZrG+qnfEqVtRRZqzzDJzWjuUqdLUCV+t6wmkJRt3vQrufVWYzW7+qJ5hbbJhy79dHC\nct3ECwBk1ugnrepzdBOvAIAJo4GOiD5vUNau73NNif5cXJOwoXh3daizGc36VUcAgNGDn7SOJa/Q\nzgewBgCccy+KyNF9XusAMM851/vByET8qbdfgXmiJSIajGjyCm0RgL7LJrpFJOycizrnYgCqAUBE\n/glAvnPuiUQNstASUVqIJu/0rkb845UPYefcR433jOH+BMAMAOdoGmShJaK0kMQn2vUAFgN4QETm\nAti01+u3A2hzzi3RNpiShba0Wzfm2ZYx1lsfmg3Dv9Y9Pdqx5Ql1m9Vt+twFXrz1MXW2crr+lK2g\nTFlVKydJ85WnrvXiXq/kMl5f15/VABaKyPqePy/rWWmQD+BlAMsAPCMiaxHfEHizc+6h/hr0NxlW\npi9y4Tb9zqloof7CxfF7XlVnAaC2/EhTXqumTD9p1e5p5UPd1PnqbOk25QoQANF2/QWKVll7tqqz\nVUX6yaIDG99VZ7dm2o77s9hRcJB6UjXDsBMR0P9gbYjof5xVRfVb5UuLKtRZAEjG/s1k7XLtGYf9\nxl7/ue8Xo7lupuQTLRHR3pI4dJB0LLRElBaSuLwr6VhoiSgtsNCmiIL1d6qzzfOXe+wJ0eB89SH9\nqWp3fOYwjz0ZOpGReDlj7agp+rD+iiMU5htmdovHoalF98G3FFkAmPT319TZrslHJw71KNuyTp2t\nn667bwoAMgBE2nU7nKLTj1O3G+psVWcBYGqu/qkjkqOfUK1ACxq6dPfJhSL67WkzsvRHNT5Xb18n\ncWhRjioXVX7uAFuRtRiTqf/eK21939Z42eCLPZ9oiYg842QYEZFnvi4xSAYWWiJKC0ncsJB03grt\nzRv0V85cuWCKoWXbR7N0+9OqXJep1dS0uVF3+tLxhsOtgmRnp24cM0g7svR3hvm5ygbQ73wDgAqP\nJ+INFocOkii8cY0+m62fZcuZdaKpHzUP3aPOdlw0R50dYzi2bktNmzrbZfhmiqz9b3U2XKLfqQcA\n9xToP85nzChXZ6uaDT8qDTsRt2RPUWfL8owTRoYia3HHebPU2epWw7mVBrvybPen6W576x8nw4iI\nPBuRy7uIiIYSn2iJiDwLcqEN+VoSsbOuRd1w+Uv3qdvNmnSwOrujQj82uq1WP94JALNLDLssAmBL\nU6c6OzevXp19J2Q7pQkAypUnUW05W3+kIgDMfFg3fm+Z/Bmdm2Hqg+WurqJc26+6jW268d/ibP2V\nM4jovy6qf3atvl0A2ZfdoM5WVBQOepZtxuWr1Z/YbbcsHdJZPT7RElFa4DpaIiLPgjx0wEJLRGmB\n62hJ7frnbIdxXDXPtl4xnZXuWJ84BKB67DzPPfGjaJTyYJmArHK6a/OH6uyVJw3+4qJot368eah5\nK7TZ3fqEyd2+AAACCUlEQVTPdtOc87z0QXtNCABc++Abprb/eLH+RK4g2N2snyBZ3ZyLT43VfeG/\n9EFD4tBeTptYrMppJ7cAfZEFgNF3f1+dBYDQRT8w5bW0k1uAocgCiIUz0diuncTTnR4G2Ca3LEU2\nWWJR/cdoqPGJlojSQizCQktE5BWfaImIPBuRhTY339Z0W4vucItW47XLlnFaixLUqbM12frDUays\nH+cgyFJuAsjb+Ii+0byiAfYmsd1d+vmGsVn6cdfiTP1V7THLNSQAyuteV2drSnWH0JRUPmnoweGG\nbHKMyEKbaladN9v4Hs1e+vGjA5w62zn7DHV2SdOzpn40jD1dlTv1oFJTu9mejtn7YOIJ6uzkpfqD\nEt/IlYF0J+ky3n5BnQ3nD34Gf7Aur/q98T2uGvTfOSJXHRARDaUon2iJiPzi0AERkWcstEREngV5\nHa23YxKrq5uCu/GYSOnKv2wz5VctnKHOFr58r6ntpqPON+VTSTKOSSw7baW65tSuWcFjEomIrDh0\nQETkWbQ7uHdZs9ASUVrgE20KKO3YbsrX5STjgmRKN8VZ+iuRAnKaYdpgoSVKUZbJLUuRBYDw3CVo\n6Mq1don2gxsWiIg8C/LyLhZaIkoLHDogIvKMhZaIyLMgn97lbWcYERHF6U8pJiKiAWGhJSLyjIWW\niMgzFloiIs9YaImIPGOhJSLy7H8BkwP25hb1ruMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f738ebdbdd0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.heatmap(t.ix[tig[tig.expressed_allele == 'Maternal'].index], xticklabels=[], yticklabels=[], cmap='RdBu_r')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
